<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html lang="en">

<!-- ~~~~~~~~~~~~~~~ HTML head data ~~~~~~~~~~~~~~~~~ -->
<head>
<meta http-equiv="Content-Type" content="text/html; charset=ISO-8859-1">
<title>SLiMSuite updates</title>

<link rel="stylesheet" href="http://www.southampton.ac.uk/~re1u06/rje_tabber.css" TYPE="text/css" MEDIA="screen">
<link rel="stylesheet" href="http://www.southampton.ac.uk/~re1u06/re1u06.css" TYPE="text/css" MEDIA="screen">

<!-- ~~~~~~~~~~~ Tabber Javascript ~~~~~~~~~~~ -->
<script type="text/javascript">
document.write('<style type="text/css">.tabber{display:none;}<\/style>');
var tabberOptions = {
 /* Optional: start manually (run tabber) at end of file
*/
'manualStartup':true
};
</script>
<!-- Load the tabber code -->
<script type="text/javascript" src="http://www.southampton.ac.uk/~re1u06/tabber.js"></script>

</head>
<!-- ~~~~~~~~~~~~~~~ End of HTML head data ~~~~~~~~~~~~~~~~~ -->

<body>
<h1>SLiMSuite updates</h1>
<h2>Updates in extras/:</h2>

</p>
<h2>Updates in legacy/:</h2>

</p>
<h2>Updates in libraries/:</h2>

</p>
<p><b>&bull; rje:</b> <i>Updated from Version 4.14.0.</i>
<br>&rarr; Version 4.14.1: Fixed matchExp method to be able to handline multilines. (Shame re.DOTALL doesn't work!)
<br>&rarr; Version 4.14.2: Modified integer commands to read/convert floats.
<br>&rarr; Version 4.15.0: Added intList() and numList() functions.
</p>
<p><b>&bull; rje_db:</b> <i>Updated from Version 1.7.5.</i>
<br>&rarr; Version 1.7.6: Added table.opt['Formatted'] = Whether table data has been successfully formatted using self.dataFormat()
<br>&rarr; Version 1.7.7: Added option to constrain table splitting to certain field values.
<br>&rarr; Version 1.8.0: Added option to store keys as tuples for correct sorting. (Make default at some point.)
</p>
<p><b>&bull; rje_genbank:</b> <i>Updated from Version 1.3.1.</i>
<br>&rarr; Version 1.3.2: Fixed bug in reverse complement sequences with introns.
</p>
<p><b>&bull; rje_iridis:</b> <i>Updated from Version 1.10.</i>
<br>&rarr; Version 1.10.1: Attempted to fix SLiMFarmer batch run problem. (Should not be setting irun=batch!)
<br>&rarr; Version 1.10.2: Trying to clean up unknown 30s pause. Might be freemem issue?
</p>
<p><b>&bull; rje_obj:</b> <i>Updated from Version 2.1.2.</i>
<br>&rarr; Version 2.1.3: Modified integer commands to read/convert floats.
</p>
<p><b>&bull; rje_qsub:</b> <i>Updated from Version 1.6.2.</i>
<br>&rarr; Version 1.6.3: Tweaked the showstart command for katana.
</p>
<p><b>&bull; rje_samtools:</b> <i>Created/Renamed/moved.</i>
<br>&rarr; Version 0.0: Initial Compilation.
<br>&rarr; Version 0.1.0: Modified version to handle multiple loci per file. (Original was for single bacterial chromosomes.)
</p>
<p><b>&bull; rje_seqlist:</b> <i>Updated from Version 1.11.0.</i>
<br>&rarr; Version 1.12.0: Added peptides/qregion reformatting and region=X,Y.
<br>&rarr; Version 1.13.0: Added summarise=T option for generating some summary statistics for sequence data. Added minlen & maxlen.
<br>&rarr; Version 1.14.0: Added splitseq=X split output sequence file according to X (gene/species) [None]
<br>&rarr; Version 1.15.0: Added names() method.
<br>&rarr; Version 1.15.1: Fixed bug with storage and return of summary stats.
<br>&rarr; Version 1.15.2: Fixed dna2prot reformatting.
<br>&rarr; Version 1.15.3: Fixed summarise bug (n=1).
</p>
<p><b>&bull; rje_sequence:</b> <i>Updated from Version 2.4.1.</i>
<br>&rarr; Version 2.5.0: Added yeast genome renaming.
<br>&rarr; Version 2.5.1: Modified reverse complement code.
<br>&rarr; Version 2.5.2: Tried to speed up dna2prot code.
</p>
<p><b>&bull; rje_slimcalc:</b> <i>Updated from Version 0.9.</i>
<br>&rarr; Version 0.9.1: Modified combining of motif stats to handle expectString format for individual values.
<br>&rarr; Version 0.9.2: Changed default conscore in docstring to RLC.
</p>
<p><b>&bull; rje_slimcore:</b> <i>Updated from Version 2.7.3.</i>
<br>&rarr; Version 2.7.4: Fixed walltime server bug.
<br>&rarr; Version 2.7.5: Fixed feature masking.
</p>
<p><b>&bull; rje_slimlist:</b> <i>Updated from Version 1.7.2.</i>
<br>&rarr; Version 1.7.3: Fixed bug that could not accept variable length motifs from commandline. Improved error message.
</p>
<p><b>&bull; rje_taxonomy:</b> <i>Updated from Version 1.0.</i>
<br>&rarr; Version 1.1.0: Added parsing of yeast strains.
</p>
<p><b>&bull; rje_tree:</b> <i>Updated from Version 2.11.2.</i>
<br>&rarr; Version 2.12.0: Added treeLen() method.
<br>&rarr; Version 2.13.0: Updated PNG saving with R to use newer code.
</p>
<p><b>&bull; rje_uniprot:</b> <i>Updated from Version 3.21.3.</i>
<br>&rarr; Version 3.21.4: Fixed Feature masking. Should this be switched off by default?
</p>
<p><b>&bull; rje_xref:</b> <i>Updated from Version 1.6.0.</i>
<br>&rarr; Version 1.7.0: Added comments=LIST ist of comment line prefixes marking lines to ignore (throughout file) ['//','%']
<br>&rarr; Version 1.7.1: Added xreformat=T/F : Whether to apply field reformatting to input xrefdata (True) or just xrefs to map (False) [False]
<br>&rarr; Version 1.8.0: Added recognition and parsing of yeast.txt XRef file from Uniprot (http://www.uniprot.org/docs/yeast.txt)
</p>
<p><b>&bull; snp_mapper:</b> <i>Created/Renamed/moved.</i>
<br>&rarr; Version 0.0: Initial Compilation. Batch mode for mapping SNPs needs updating.
<br>&rarr; Version 0.1: SNP mapping against a GenBank file.
<br>&rarr; Version 0.2: Fixed complement strand bug.
<br>&rarr; Version 0.3.0: Updated to work with RATT(/Mummer?) snp output file. Improved docs.
<br>&rarr; Version 0.4.0: Major reworking for easier updates and added functionality. (Convert to 1.0.0 when complete.)
</p>
<h2>Updates in tools/:</h2>

</p>
<p><b>&bull; gablam:</b> <i>Updated from Version 2.19.2.</i>
<br>&rarr; Version 2.20.0: Added SNP Table output.
</p>
<p><b>&bull; gopher:</b> <i>Updated from Version 3.4.1.</i>
<br>&rarr; Version 3.4.2: Removed GOPHER System Exit on IOError to prevent breaking of REST server.
</p>
<p><b>&bull; pagsat:</b> <i>Created/Renamed/moved.</i>
<br>&rarr; Version 1.0.0: Initial working version for based on rje_pacbio assessment=T.
<br>&rarr; Version 1.1.0: Fixed bug with gene and protein summary data. Removed gene/protein reciprocal searches. Added compare mode.
<br>&rarr; Version 1.1.1: Added PAGSAT output directory for tidiness!
<br>&rarr; Version 1.1.2: Renamed the PacBio class PAGSAT.
<br>&rarr; Version 1.2.0: Tidied up output directories. Added QV filter and Top Gene/Protein hits output.
<br>&rarr; Version 1.2.1: Added casefilter=T/F  : Whether to filter leading/trailing lower case (low QV) sequences [True]
<br>&rarr; Version 1.3.0: Added tophitbuffer=X and initial synteny analysis for keeping best reference hits.
<br>&rarr; Version 1.4.0: Added chrom-v-contig alignment files along with *.ordered.fas.
<br>&rarr; Version 1.4.1: Made default chromalign=T.
<br>&rarr; Version 1.4.2: Fixed casefilter=F.
<br>&rarr; Version 1.5.0: diploid=T/F     : Whether to treat assembly as a diploid [False]
<br>&rarr; Version 1.6.0: mincontiglen=X  : Minimum contig length to retain in assembly [1000]
<br>&rarr; Version 1.6.1: Added diploid=T/F to R PNG call.
</p>
<p><b>&bull; peptcluster:</b> <i>Updated from Version 1.5.1.</i>
<br>&rarr; Version 1.5.2: Improved clarity of warning message.
</p>
<p><b>&bull; pingu_V4:</b> <i>Updated from Version 4.5.0.</i>
<br>&rarr; Version 4.5.1: Debugging missing identifiers and indexing speed. Added good and bad DB.
<br>&rarr; Version 4.5.2: Fixed SIF output and changed names to sif-* for opening in browser.
<br>&rarr; Version 4.5.3: Updated REST output.
</p>
<p><b>&bull; seqsuite:</b> <i>Updated from Version 1.8.0.</i>
<br>&rarr; Version 1.9.0: Added PAGSAT and SMRTSCAPE.
<br>&rarr; Version 1.9.1: Fixed HAQESAC setobjects=True error.
<br>&rarr; Version 1.10.0: Added batchrun=FILELIST batcharg=X batch running mode.
<br>&rarr; Version 1.11.0: Added SAMTools and Snapper/SNP_Mapper.
</p>
<p><b>&bull; slimbench:</b> <i>Updated from Version 2.10.0.</i>
<br>&rarr; Version 2.10.1: Updated ELM Source URLs.
</p>
<p><b>&bull; slimfarmer:</b> <i>Updated from Version 1.4.2.</i>
<br>&rarr; Version 1.4.3: Added recognition of missing slimsuite programs and switching to slimsuite=F.
</p>
<p><b>&bull; slimfinder:</b> <i>Updated from Version 5.2.0.</i>
<br>&rarr; Version 5.2.1: Fixed ambocc<1 and minocc<1 issue. (Using integers rather than floats.) Fixed OccRes Sig output format.
</p>
<p><b>&bull; slimparser:</b> <i>Updated from Version 0.3.1.</i>
<br>&rarr; Version 0.3.2: Fixed issue reading files for full output.
<br>&rarr; Version 0.3.3: Tidied output names when restbase=jobid.
</p>
<p><b>&bull; slimprob:</b> <i>Updated from Version 2.2.3.</i>
<br>&rarr; Version 2.2.4: Improved slimcalc output (s.f.).
</p>
<p><b>&bull; slimsuite:</b> <i>Updated from Version 1.5.0.</i>
<br>&rarr; Version 1.5.1: Changed disorder to iuscore to avoid module conflict.
</p>
<p><b>&bull; smrtscape:</b> <i>Created/Renamed/moved.</i>
<br>&rarr; Version 0.0.0: Initial Compilation.
<br>&rarr; Version 1.0.0: Initial working version for server.
<br>&rarr; Version 1.1.0: Added xnlist=LIST : Additional columns giving % sites with coverage >= Xn [10,25,50,100].
<br>&rarr; Version 1.2.0: Added assessment -> now PAGSAT.
<br>&rarr; Version 1.3.0: Added seed and anchor read coverage generator (calculate=T).
<br>&rarr; Version 1.3.1: Deleted assessment function. (Now handled by PAGSAT.)
<br>&rarr; Version 1.4.0: Added new coverage=T function that incorporates seed and anchor subreads.
<br>&rarr; Version 1.5.0: Added parseparam=FILES with paramlist=LIST to parse restricted sets of parameters.
<br>&rarr; Version 1.6.0: New SMRTSCAPE program building on PacBio v1.5.0. Added predict=T/F option.
<br>&rarr; Version 1.6.1: Updated parameters=T to incorporate that the seed read counts as X=1.
<br>&rarr; Version 1.7.0: Added *.summary.tdt output from subread summary analysis. Added minreadlen.
<br>&rarr; Version 1.8.0: preassembly=FILE: Preassembly fasta file to assess/correct over-fragmentation (use seqin=FILE for subreads)
</p>
<!-- ~~~~~~~~~~~~~~ HTML tail data ~~~~~~~~~~~~~~~~~ -->
<HR><FONT COLOR=#979E45 SIZE=2>&copy; RJ Edwards 2015. Last modified 30 Nov 2015.</FONT></P>

<script type="text/javascript">
/* manualStartup=true so need to run it now */
tabberAutomatic(tabberOptions);
</script>

</body>
</html>
<!-- ~~~~~~~~~~~~~~ End of HTML tail data ~~~~~~~~~~~~~~~~~ -->