<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html lang="en">

<!-- ~~~~~~~~~~~~~~~ HTML head data ~~~~~~~~~~~~~~~~~ -->
<head>
<meta http-equiv="Content-Type" content="text/html; charset=ISO-8859-1">
<title>SLiMSuite updates</title>

<link rel="stylesheet" href="http://www.southampton.ac.uk/~re1u06/rje_tabber.css" TYPE="text/css" MEDIA="screen">
<link rel="stylesheet" href="http://www.southampton.ac.uk/~re1u06/re1u06.css" TYPE="text/css" MEDIA="screen">

<!-- ~~~~~~~~~~~ Tabber Javascript ~~~~~~~~~~~ -->
<script type="text/javascript">
document.write('<style type="text/css">.tabber{display:none;}<\/style>');
var tabberOptions = {
 /* Optional: start manually (run tabber) at end of file
*/
'manualStartup':true
};
</script>
<!-- Load the tabber code -->
<script type="text/javascript" src="http://www.southampton.ac.uk/~re1u06/tabber.js"></script>

</head>
<!-- ~~~~~~~~~~~~~~~ End of HTML head data ~~~~~~~~~~~~~~~~~ -->

<body>
<h1>SLiMSuite updates</h1>
<h2>Updates in extras/:</h2>

</p>
<p><b>&bull; rje_pydocs:</b> <i>Updated from Version 2.16.2.</i>
<br>&rarr; Version 2.16.3: Fixed docstring REST parsing to work with _V* modules.
</p>
<h2>Updates in legacy/:</h2>

</p>
<h2>Updates in libraries/:</h2>

</p>
<p><b>&bull; rje:</b> <i>Updated from Version 4.15.1.</i>
<br>&rarr; Version 4.16.0: Added list2dict(inlist,inkeys) and dict2list(indict,inkeys) functions.
<br>&rarr; Version 4.16.1: Improved handling of integer parameters when given bad commands.
<br>&rarr; Version 4.17.0: Added extra functions to randomList()
</p>
<p><b>&bull; rje_blast_V2:</b> <i>Updated from Version 2.9.1.</i>
<br>&rarr; Version 2.10.0: Added nocoverage calculation based on local alignment table.
<br>&rarr; Version 2.11.0: Added localFragFas output method.
<br>&rarr; Version 2.11.1: Fixed snp local table revcomp bug. [Check this!]
<br>&rarr; Version 2.11.2: Fixed GABLAM calculation bug when '*' in protein sequences.
</p>
<p><b>&bull; rje_db:</b> <i>Updated from Version 1.8.0.</i>
<br>&rarr; Version 1.8.1: Added sfdict to saveTable output.
</p>
<p><b>&bull; rje_genbank:</b> <i>Updated from Version 1.3.2.</i>
<br>&rarr; Version 1.4.0: Added addtags=T/F : Add locus_tag identifiers if missing - needed for gene/cds/prot fasta output [False]
<br>&rarr; Version 1.4.1: Fixed genetic code warning.
<br>&rarr; Version 1.5.0: Added setupRefGenome() method based on PAGSAT code.
<br>&rarr; Version 1.5.1: Fixed logskip append locus sequence file bug.
<br>&rarr; Version 1.5.2: Fixed addtag(s) bug.
</p>
<p><b>&bull; rje_hprd:</b> <i>Updated from Version 1.2.</i>
<br>&rarr; Version 1.2.1: Fixed "PROTEIN_ARCHITECTURE" bug.
</p>
<p><b>&bull; rje_menu:</b> <i>Updated from Version 0.3.</i>
<br>&rarr; Version 0.4.0: Changed handling of default for exiting menu loop. May affect behaviour of some existing menus.
</p>
<p><b>&bull; rje_mitab:</b> <i>Updated from Version 0.2.0.</i>
<br>&rarr; Version 0.2.1: Fixed redundant evidence/itype bug (primarily dip)
</p>
<p><b>&bull; rje_obj:</b> <i>Updated from Version 2.1.3.</i>
<br>&rarr; Version 2.2.0: Added screenwrap=X.
<br>&rarr; Version 2.2.1: Improved handling of integer parameters when given bad commands.
</p>
<p><b>&bull; rje_samtools:</b> <i>Updated from Version 0.1.0.</i>
<br>&rarr; Version 0.2.0: Added majmut=T/F : Whether to restrict output and stats to positions with non-reference Major Allele [False]
<br>&rarr; Version 1.0.0: Major reworking. Old version frozen as rje_samtools_V0.
<br>&rarr; Version 1.1.0: Added snptabmap=X,Y alternative SNPTable mapping and read_depth statistics []. Added majref=T/F.
<br>&rarr; Version 1.2.0: Added developmental combining of read mapping onto two different genomes.
<br>&rarr; Version 1.3.0: Major debugging and code clean up.
<br>&rarr; Version 1.4.0: Added parsing of read number (to link SNPs) and fixed deletion error at same time. Added rid=T/F and snponly=T/F.
<br>&rarr; Version 1.5.0: Added biallelic=T/F   : Whether to restrict SNPs to pure biallelic SNPs (two alleles meeting mincut) [False]
<br>&rarr; Version 1.5.1: Fixed REF/Ref ALT/Alt bug.
<br>&rarr; Version 1.6.0: Added majfocus=T/F : Whether the focus is on Major Alleles (True) or Mutant/Reference Alleles (False) [True]
<br>&rarr; Version 1.7.0: Added parsing of *.sam files for generating RID table.
<br>&rarr; Version 1.8.0: Added read coverage summary/checks.
<br>&rarr; Version 1.8.1: Fixed issue when RID file not generated by pileup parsing. Set RID=True by default to avoid issues.
</p>
<p><b>&bull; rje_samtools_V0:</b> <i>Created/Renamed/moved.</i>
<br>&rarr; Version 0.0: Initial Compilation.
<br>&rarr; Version 0.1.0: Modified version to handle multiple loci per file. (Original was for single bacterial chromosomes.)
<br>&rarr; Version 0.2.0: Added majmut=T/F : Whether to restrict output and stats to positions with non-reference Major Allele [False]
</p>
<p><b>&bull; rje_seq:</b> <i>Updated from Version 3.23.0.</i>
<br>&rarr; Version 3.24.0: Added REST seqout output.
</p>
<p><b>&bull; rje_seqlist:</b> <i>Updated from Version 1.15.3.</i>
<br>&rarr; Version 1.15.4: Fixed REST server output bug.
<br>&rarr; Version 1.15.5: Fixed reformat=fasta default issue introduced from fixing REST output bug.
<br>&rarr; Version 1.16.0: Added edit=T sequence edit mode upon loading (will switch seqmode=list).
<br>&rarr; Version 1.17.0: Added additional summarise=T output for seqmode=db.
<br>&rarr; Version 1.18.0: Added revcomp to reformat options.
<br>&rarr; Version 1.19.0: Added option log description for deleting sequence during edit.
<br>&rarr; Version 1.20.0: Added option to give a file of changes for edit mode.
<br>&rarr; Version 1.20.1: Fixed edit=FILE deletion bug.
</p>
<p><b>&bull; rje_sequence:</b> <i>Updated from Version 2.5.2.</i>
<br>&rarr; Version 2.5.3: Fixed genetic code warning error.
</p>
<p><b>&bull; rje_slimcore:</b> <i>Updated from Version 2.7.5.</i>
<br>&rarr; Version 2.7.6: Added feature masking log info or warning.
<br>&rarr; Version 2.7.7: Switched feature masking OFF by default to give consistent Uniprot versus FASTA behaviour.
</p>
<p><b>&bull; rje_synteny:</b> <i>Created/Renamed/moved.</i>
<br>&rarr; Version 0.0.0: Initial Compilation.
</p>
<p><b>&bull; rje_taxonomy:</b> <i>Updated from Version 1.1.0.</i>
<br>&rarr; Version 1.2.0: Added storage of Parents.
</p>
<p><b>&bull; rje_tree:</b> <i>Updated from Version 2.13.0.</i>
<br>&rarr; Version 2.14.0: Added cladeSpec().
</p>
<p><b>&bull; rje_uniprot:</b> <i>Updated from Version 3.21.4.</i>
<br>&rarr; Version 3.22.0: Tweaked REST table output.
</p>
<p><b>&bull; rje_xref:</b> <i>Updated from Version 1.8.0.</i>
<br>&rarr; Version 1.8.1: Added rest run mode to avoid XRef table output if no gene ID list is given. Added `genes` and `genelist` as `idlist=LIST` synonym.
<br>&rarr; Version 1.8.2: Catching self.dict['Mapping'] error for REST server.
</p>
<p><b>&bull; snp_mapper:</b> <i>Updated from Version 0.4.0.</i>
<br>&rarr; Version 0.5.0: Added CDS rating.
<br>&rarr; Version 0.6.0: Added AltFT mapping mode (map features to AltLocus and AltPos)
<br>&rarr; Version 0.7.0: Added additional fields for processing Snapper output. (Hopefully will still work for SAMTools etc.)
<br>&rarr; Version 0.8.0: Added parsing of GFF file from Prokka.
<br>&rarr; Version 0.8.1: Corrected "intron" classification for first position of features. Updated FTBest defaults.
<br>&rarr; Version 1.0.0: Version that works with Snapper V1.0.0. Not really designed for standalone running any more.
</p>
<h2>Updates in tools/:</h2>

</p>
<p><b>&bull; comparimotif_V3:</b> <i>Updated from Version 3.12.</i>
<br>&rarr; Version 3.13.0: Added REST server function.
</p>
<p><b>&bull; gablam:</b> <i>Updated from Version 2.20.0.</i>
<br>&rarr; Version 2.21.0: Added nocoverage Table output of regions missing from pairwise SNP Table.
<br>&rarr; Version 2.21.1: Added fragrevcomp=T/F : Whether to reverse-complement DNA fragments that are on reverse strand to query [True]
<br>&rarr; Version 2.22.0: Added description to HitSum table.
<br>&rarr; Version 2.22.1: Added localaln=T/F to keep local alignment sequences in the BLAST local Table.
<br>&rarr; Version 2.22.2: Fixed local output error. (Query/Qry issue - need to fix this and make consistent!)
<br>&rarr; Version 2.22.3: Fixed blastv and blastb error: limit also applies to individual pairwise hits!
<br>&rarr; Version 2.23.0: Divided GablamFrag and FragMerge.
</p>
<p><b>&bull; pagsat:</b> <i>Updated from Version 1.6.1.</i>
<br>&rarr; Version 1.7.0: Added tidy=T/F option. (Development)
<br>&rarr; Version 1.7.1: Updated tidy=T/F to include initial assembly.
<br>&rarr; Version 1.7.2: Fixed some bugs introduced by changing gablam fragment output.
<br>&rarr; Version 1.7.3: Added circularise sequence generation.
<br>&rarr; Version 1.8.0: Added orphan processing and non-chr naming of Reference.
<br>&rarr; Version 1.9.0: Modified the join sorting and merging. Added better tracking of positions when trimming.
<br>&rarr; Version 1.9.1: Added joinmargin=X    : Number of extra bases allowed to still be considered an end local BLAST hit [10]
<br>&rarr; Version 1.10.0: Added weighted tree output and removed report warning.
<br>&rarr; Version 1.10.1: Fixed issue related to having Description in GABLAM HitSum tables.
<br>&rarr; Version 1.10.2: Tweaked haploid core output.
<br>&rarr; Version 1.10.3: Fixed tidy bug for RevComp contigs and switched joinsort default to Identity. (Needs testing.)
<br>&rarr; Version 1.10.4: Added genetar option to tidy out genesummary and protsummary output. Incorporated rje_synteny.
<br>&rarr; Version 1.10.5: Set gablamfrag=1 for gene/protein hits.
<br>&rarr; Version 1.11.0: Consolidated automated tidy mode and cleaned up some excess code.
<br>&rarr; Version 1.11.1: Added option for running self-PAGSAT of ctidX contigs versus haploid set. Replaced ctid "X" with "N".
<br>&rarr; Version 1.11.2: Fixed Snapper run choice bug.
</p>
<p><b>&bull; pingu_V4:</b> <i>Updated from Version 4.5.3.</i>
<br>&rarr; Version 4.6.0: Added hubonly=T/F : Whether to restrict pairwise PPI to those with both hub and spoke in hublist [False]
<br>&rarr; Version 4.6.1: Fixed some ppifas=T/F bugs and added combineppi=T/F : Whether to combine all spokes into a single fasta file [False]
<br>&rarr; Version 4.6.2: Added check/filter for multiple SpokeUni pointing to same sequence. (Compilation redundancy mapping failure!)
<br>&rarr; Version 4.6.3: Fixed issue with 1:many SpokeUni:Spoke mappings messing up XHub.
<br>&rarr; Version 4.7.0: Added ppidbreport=T/F : Summary output for PPI compilation of evidence/PPIType/DB overlaps [True]
<br>&rarr; Version 4.8.0: Fixed report duplication issue and added additional summary output
</p>
<p><b>&bull; qslimfinder:</b> <i>Updated from Version 2.1.0.</i>
<br>&rarr; Version 2.1.1: Switched feature masking OFF by default to give consistent Uniprot versus FASTA behaviour.
</p>
<p><b>&bull; seqsuite:</b> <i>Updated from Version 1.11.0.</i>
<br>&rarr; Version 1.11.1: Redirected PacBio to call SMRTSCAPE.
<br>&rarr; Version 1.11.2: Fixed batchrun batchlog=False log error.
<br>&rarr; Version 1.12.0: Added Snapper.
</p>
<p><b>&bull; slimfarmer:</b> <i>Updated from Version 1.4.3.</i>
<br>&rarr; Version 1.4.4: Modified default vmem request to 126GB from 127GB.
<br>&rarr; Version 1.4.5: Updated BLAST loading default to 2.2.31
</p>
<p><b>&bull; slimfinder:</b> <i>Updated from Version 5.2.1.</i>
<br>&rarr; Version 5.2.2: Added warnings for ambocc and minocc that exceed the absolute minima. Updated docstring.
<br>&rarr; Version 5.2.3: Switched feature masking OFF by default to give consistent Uniprot versus FASTA behaviour. Fixed FTMask=T/F bug.
</p>
<p><b>&bull; slimparser:</b> <i>Updated from Version 0.3.3.</i>
<br>&rarr; Version 0.3.4: Tweaked error messages.
<br>&rarr; Version 0.4.0: Added simple json format output.
</p>
<p><b>&bull; slimprob:</b> <i>Updated from Version 2.2.4.</i>
<br>&rarr; Version 2.2.5: Fixed FTMask=T/F bug.
</p>
<p><b>&bull; slimsearch:</b> <i>Updated from Version 1.7.</i>
<br>&rarr; Version 1.7.1: Minor modification to docstring. Preparation for update to SLiMSearch 2.0 optimised for proteome searches.
</p>
<p><b>&bull; slimsuite:</b> <i>Updated from Version 1.5.1.</i>
<br>&rarr; Version 1.5.2: Updated XRef REST call.
<br>&rarr; Version 1.6.0: Removed SLiMCore as default. Default will now show help.
</p>
<p><b>&bull; smrtscape:</b> <i>Updated from Version 1.8.0.</i>
<br>&rarr; Version 1.9.0: Updated empirical preassembly mapefficiency calculation.
<br>&rarr; Version 1.10.0: Added batch processing of subread files.
<br>&rarr; Version 1.10.1: Fixed bug in batch processing.
</p>
<p><b>&bull; snapper:</b> <i>Created/Renamed/moved.</i>
<br>&rarr; Version 0.0.0: Initial Compilation.
<br>&rarr; Version 0.1.0: Tidied up with improved run pickup.
<br>&rarr; Version 0.2.0: Added FASTQ and improved CNV output along with all features.
<br>&rarr; Version 0.2.1: Fixed local output error. (Query/Qry issue - need to fix this and make consistent!) Fixed snp local table revcomp bug.
<br>&rarr; Version 0.2.2: Corrected excess CNV table output (accnum AND shortname).
<br>&rarr; Version 0.2.3: Corrected "intron" classification for first position of features. Updated FTBest defaults.
<br>&rarr; Version 1.0.0: Working version with completed draft manual. Added to SeqSuite.
<br>&rarr; Version 1.0.1: Fixed issues when features missing.
</p>
<!-- ~~~~~~~~~~~~~~ HTML tail data ~~~~~~~~~~~~~~~~~ -->
<HR><FONT COLOR=#979E45 SIZE=2>&copy; RJ Edwards 2016. Last modified 12 Sep 2016.</FONT></P>

<script type="text/javascript">
/* manualStartup=true so need to run it now */
tabberAutomatic(tabberOptions);
</script>

</body>
</html>
<!-- ~~~~~~~~~~~~~~ End of HTML tail data ~~~~~~~~~~~~~~~~~ -->