###################################################
### SNP Frequency Plotting Tool - R Graphics ~~ ###
### VERSION: 0.5.0                        ~~~~~ ###
### LAST EDIT: 24/11/17                   ~~~~~ ###
### AUTHORS: Richard Edwards 2017         ~~~~~ ###
### CONTACT: richard.edwards@unsw.edu.au  ~~~~~ ###
###################################################

################# ::: HISTORY ::: ######################
# v0.0.0 : Initial version based on PAGSAT.R v0.7.3.
# v0.1.0 : Modifying to run from commandline. Deleted excess PAGSAT code. Renamed from Snapper.R.
# v0.2.0 : Added fdrplot=T/F as an option. Fixed CNV plot zoom bug.
# v0.3.0 : Added multiplot=T/F as an option.
# v0.4.0 : Added feature plots for reference and fixed FDR plotting.
# v0.5.0 : Updated for plots lacking Snapper data.


#!# Rearrange a little and make more elements into functions (load data etc.)
#!# Move functions shared with PAGSAT into a separate file
#!# Add loading of regions to mark from separate table
#!# Plot 0CNV regions on the SNPFreq plots

############### ::: GENERAL SETUP ::: ##################
# Usage = Rscript rje.r snpfreq basefile [options] 
# Main options: snapbase=X    : base path to Snapper files
#               freqpath=X    : path to Pileup frequency outputs
#               freqcomp=LIST : list of comparisons to make
#               chrom=X, xmin=X, xmax=X : Limit plots to a specific chromosome and/or region
#               pcut=X        : Minimum p-value to shade in yellow [0.01]

#i# As this is the Snapper.R script, basefile should refer to the base of the Snapper run.
#i# |-- V0.1.0 uses the *.cnv.tdt table only. (This might be for a custom plot.)
#i# NOTE: This is a change from V0.0.0, in which this was assigned to snpbase and basefile was used for the SNP frequency runs

#># SNP Frequency run data is now set by freqpath=X + freqcomp=LIST
#># |-- *.fdr.tdt
#># |-- *.combinedsnp.tdt

#i# NOTE: Multipanel SNP Frequency plots will need to take multiple freqcomp inputs: use the number to define the number of panels

#># Output file names will be based on a outbase=X, which is now given by basefile=X
#># Input for Snapper files (Snapper basefile) is now given by snapbase=X

setTesting = function(){
  arglist = list("rtype"="snpfreq","rdir"="/Users/redwards/Dropbox/_Repository_/slimsuite/libraries/r/")
  arglist$snapbase = "/Volumes/bioinf-1/aperezbercoff/Microbiogen-Nov14/dev/2016-06-21.PileupSNPMap/MBG11A.MBG16533.vs.sgd.ref"
  arglist$freqpath = "/Volumes/bioinf-1/aperezbercoff/Microbiogen-Nov14/dev/2016-06-21.PileupSNPMap/"
  arglist$freqcomp = "Gal.vs.Glu.biallelic,Xyl.vs.Glu.biallelic,P12.vs.P6.biallelic"
  arglist$basefile = "Gal.vs.Glu.test"
  arglist$refbase = "/Users/redwards/ownCloud/projects/MBGHybridSelection-Feb17/data/2017-03-10.Ref.S288C/sgd.R64.2.1"
  setwd("/Users/redwards/ownCloud/projects/Microbiogen-Nov14/dev/2017-03-20.SNPFreqTest")
}

## ~ Read commandline arguments ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ##
rtype = arglist$rtype        # arglist is generated by rje.r
basefile = arglist$basefile  # arglist is generated by rje.r
#print(basefile)

## ~ Set up global parameters and file names ~~~~~~~~~~~~~~~~~~~~~~~~~~ ##
settings = list(basefile=basefile)  # IO Basefile Access with settings[["basefile"]] or settings$basefile
settings$units = "kb"
settings$pngwidth = 2400
settings$pngheight = 1600
settings$dualpng = 1600
settings$pointsize = 36
settings$pcut = 0.01
settings$fdrplot = FALSE
settings$multiplot = FALSE

## ~ Setup Chromosome region settings ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ##
settings$chrom = NA
settings$ymin=0
settings$ymax=1.0
settings$xmin=0
settings$xmax=-1

## ~ Setup chromosome plot and features methods ~~~~~~~~~~~~~~~~~~~~~~~~ ##
rjesource("rje_genomics.r")   
settings = featureDefaults(settings)

## ~ Update default settings from commandline arguments (setting=value) ~ ##
(arglist)  # This contains settings read in from arguments
arglist = argBool(arglist,c("assessment","covplot","diploid","genesummary","protsummary","chromalign","fdrplot","multiplot"))    # List of Boolean settings
arglist = argNum(arglist,c("pngwidth","pngheight","pointsize","minloclen","topcontigs","pcut","ymin","ymax","xmin","xmax"))     # List of numeric settings
arglist = argList(arglist,c("plotft","freqcomp"))
for(cmd in names(arglist)){
  settings[[cmd]] = arglist[[cmd]]  # Need to add processing of booleans etc.
}
(settings)  # This contains settings read in from arguments

# Setup base filenames for outputs
settings$outbase = settings$basefile
(settings$basename = settings$outbase)          # Output basefile without path
(settings$plotdir = paste(settings$outbase,".Plots/",sep=""))         # Directory for Plot output
(settings$pngbase = paste(settings$plotdir,settings$basename,sep=""))     # Directory for Plot output
if(file.exists(settings$plotdir)==FALSE){
  dir.create(settings$plotdir)
}

# Setup scaling and units
settings$scaling = 1
if(settings$units == "kb"){
  settings$scaling = 1000
}
if(settings$units == "Mb"){
  settings$scaling = 1e6
}

# Setup colour profile
if(settings$features){
  writeLines("Features for plotting:")
  writeLines(settings$plotft)
}
settings = featureColours(settings)
settings$testing=FALSE

#print(settings)  # This contains final settings


############# ::: LOAD AND TIDY DATA ::: ##################

## Load the SNP frequency data for each input comparison
fdrdb = list()  # Each input freqcomp will have an entry
combdb = list()
for(freqcomp in settings$freqcomp){
  fdrdb[[freqcomp]] = read.table( paste(settings$freqpath,freqcomp,".fdr.tdt",sep=""), header = T, stringsAsFactors = T, sep = '\t', quote = '')
  summary(fdrdb[[freqcomp]])
  if(! 'AltPos' %in% colnames(fdrdb[[freqcomp]])){
      fdrdb[[freqcomp]]$AltPos = fdrdb[[freqcomp]]$Pos
      fdrdb[[freqcomp]]$AltLocus = fdrdb[[freqcomp]]$Locus
  }

  if(file.exists(paste(settings$freqpath,freqcomp,".combinedsnp.tdt",sep=""))){
      freqfile = paste(settings$freqpath,freqcomp,".combinedsnp.tdt",sep="")
  }else{
      freqfile = paste(settings$freqpath,freqcomp,".snpcomb.tdt",sep="")
  }
  combdb[[freqcomp]] = read.table( freqfile, header = T, stringsAsFactors = T, sep = '\t', quote = '')
  summary(combdb[[freqcomp]])
  if(! 'AltPos' %in% colnames(combdb[[freqcomp]])){
      combdb[[freqcomp]]$AltPos = combdb[[freqcomp]]$Pos
      combdb[[freqcomp]]$AltLocus = combdb[[freqcomp]]$Locus
  }
}


#### >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> ####




# Update colour dicionary for contigs
cx = 0
for(chrom in altchr){
  cx = cx + 1
  if(cx > length(soton$col)){ cx = 1 }
  settings$col[[chrom]] = soton$col[cx]
  cname = strsplit(chrom,"_")[[1]][1]
  settings$col[[cname]] = soton$col[cx]
}
# Update colour dicionary for chromosomes
cx = 0
for(chrom in refchr){
  cx = cx + 1
  if(cx > length(soton$col)){ cx = 1 }
  settings$col[[chrom]] = soton$col[cx]
  cname = strsplit(chrom,"_")[[1]][1]
  settings$col[[cname]] = soton$col[cx]
}

## ~ Load features data ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ##
if(is.null(settings$refbase)){
  if(settings$multiplot == FALSE){ writeLines("No RefBase set - no feature plots") }
  ftdb = list("Ref"=data.frame())
}else{
  ftdata = featureSetup(settings$refbase,ftype="Ref")
  print(table(ftdata$feature))
  ftdb = list("Ref"=ftdata)
}

# Set test data
if(settings$testing){
  chrom = altchr[3]
  chromdata = cnvdata
  chromCNVPlot(chromdata,chrom)
  chromCNVPlot(chromdata,refchr[3])
}


# Set test data
if(settings$testing){
  chrom = altchr[3]
  chromdata = cnvdata
  alt="Alt"
  (comparison = settings$freqcomp[1])
  snpdata = fdrdb[[comparison]]
  snpdata = combdb[[comparison]]
  chromSNPFreqPlot(snpdata,chromdata,chrom,alt="Alt",comparison=comparison,ymin=0,ymax=1.0,xmin=0,xmax=-1)
  chromSNPFreqPlot(snpdata,chromdata,chrom,alt="Alt",comparison=comparison,ymin=0,ymax=1.0,xmin=0,xmax=-1,colbychrom = FALSE)
  chromSNPFreqExpPlot(snpdata,chromdata,chrom,alt="Alt",comparison=comparison,ymin=0,ymax=1.0,xmin=0,xmax=-1)
  chromSNPFreqExpPlot(snpdata,chromdata,chrom,alt="Alt",comparison=comparison,ymin=0,ymax=1.0,xmin=0,xmax=-1,colbychrom = FALSE)
}

if(settings$testing){
  chrom = refchr[3]
  multiSNPCNVPlot("pngfile",chrom,alt="Ref",ymin=0,ymax=1.0,xmin=0,xmax=-1,makepng=FALSE)
  chrom = altchr[3]
  multiSNPCNVPlot("pngfile",chrom,alt="Alt",ymin=0,ymax=1.0,xmin=0,xmax=-1,makepng=FALSE)
}
if(settings$testing){
  settings$pcut = 0.001  # Controls which points are shaded yellow
  settings$fdrplot = TRUE
  multiSNPPlot("pngfile",chrom,alt="Ref",ymin=0,ymax=1.0,xmin=0,xmax=-1,makepng=FALSE)
}


if(settings$testing){
  multiPlotFull()
}

### Generate PNG Plots
#!# Add settings to specificy which plots to make and enable zooming in

zoom = settings$chrom %in% refchr ||  settings$chrom %in% altchr

if(zoom){
  writeLines(c("","Zoom plots..."))
  zoomPlot()
}else{
  if(settings$multiplot){
    writeLines(c("","Multichromosome plots..."))
    multiPlotFull()
  }else{
    writeLines(c("","Chromosome plots..."))
    soloPlots()
  }
}
  
writeLines(">>Finished")





### !!! ### OLD PLOTS TO DELETE ONCE HAPPY IT'S WORKING !!! ###

#!# Modified from above. Possibly make more flexible with data frames as input
multiSNPPlot = function(pngfile,chrom,alt="Alt",ymin=0,ymax=1.0,xmin=0,xmax=-1,colbychrom=FALSE,region=FALSE){
  png(filename=pngfile, width=settings$pngwidth, height=settings$pngheight, units = "px", pointsize=settings$pointsize)
  par(mar=c(2,6,2,1))
  panels = c(1,2,3)
  layout(matrix(panels,byrow=FALSE,nrow=3))
  chromSNPFreqPlot(galdata,cnvdata,chrom,alt,"Galactose vs Glucose",ymin,ymax,xmin,xmax,colbychrom,region)
  chromSNPFreqPlot(xyldata,cnvdata,chrom,alt,"Xylose vs Glucose",ymin,ymax,xmin,xmax,colbychrom,region)
  par(mar=c(5,6,2,1))
  chromSNPFreqPlot(popdata,cnvdata,chrom,alt,"Evolved Population P12 vs P6",0.0,ymax,xmin,xmax,colbychrom,region)
  #NB. Dropped chromCNVPlot(cnvdata,chrom) as fourth plot: not that informative in this context.
  dev.off()
}

multiSNPExpPlot = function(pngfile,chrom,alt="Alt",ymin=0,ymax=1.0,xmin=0,xmax=-1,colbychrom=FALSE,region=FALSE){
  png(filename=pngfile, width=settings$pngwidth, height=settings$pngheight, units = "px", pointsize=settings$pointsize)
  par(mar=c(2,6,2,1))
  panels = c(1,2,3)
  layout(matrix(panels,byrow=FALSE,nrow=3))
  chromSNPFreqExpPlot(galdata,cnvdata,chrom,alt,"Galactose",ymin,ymax,xmin,xmax,colbychrom,region)
  chromSNPFreqExpPlot(xyldata,cnvdata,chrom,alt,"Xylose",ymin,ymax,xmin,xmax,colbychrom,region)
  #par(mar=c(5,6,2,1))
  chromSNPFreqExpPlot(xyldata,cnvdata,chrom,alt,"Glucose",ymin,ymax,xmin,xmax,colbychrom,region,inverse=TRUE)
  #NB. Dropped chromCNVPlot(cnvdata,chrom) as fourth plot: not that informative in this context.
  dev.off()
}


makeMBGPlots = function(fileset="biallelic-11a"){
  
  (settings$basename = fileset)
  (basefile = paste("/Users/redwards/Documents/Projects/Microbiogen-Nov14/dev/2016-06-21.PileupSNPMap/",fileset,sep=""))
  (settings$pngbase = paste(basefile,".Plots/",settings$basename,sep=""))     # Directory for Plot output
  if(file.exists(paste(basefile,".Plots/",sep=""))==FALSE){
    dir.create(paste(basefile,".Plots/",sep=""))
  }
  
  (galbase = paste("/Users/redwards/Documents/Projects/Microbiogen-Nov14/dev/2016-06-21.PileupSNPMap/Gal.vs.Glu",fileset,sep="."))
  (xylbase = paste("/Users/redwards/Documents/Projects/Microbiogen-Nov14/dev/2016-06-21.PileupSNPMap/Xyl.vs.Glu",fileset,sep="."))
  (popbase = paste("/Users/redwards/Documents/Projects/Microbiogen-Nov14/dev/2016-06-21.PileupSNPMap/P12.vs.P6",fileset,sep="."))
  
  galdata = read.table( paste(galbase,".combinedsnp.tdt",sep=""), header = T, stringsAsFactors = T, sep = '\t', quote = '')
  summary(galdata)
  xyldata = read.table( paste(xylbase,".combinedsnp.tdt",sep=""), header = T, stringsAsFactors = T, sep = '\t', quote = '')
  summary(xyldata)
  popdata = read.table( paste(popbase,".combinedsnp.tdt",sep=""), header = T, stringsAsFactors = T, sep = '\t', quote = '')
  summary(popdata)
  
  settings$pcut = 0.01
  multiSNPPlot(paste(settings$pngbase,"new.exampleXV","png",sep="."),altchr[16],alt="Alt",ymin=0.0,ymax=1.0,xmin=0,xmax=400,region="Example Region:")
  multiSNPExpPlot(paste(settings$pngbase,"exp.exampleXV","png",sep="."),altchr[16],alt="Alt",ymin=0.0,ymax=1.0,xmin=0,xmax=400,region="Example Region:")
  multiSNPPlot(paste(settings$pngbase,"exampleXV","png",sep="."),altchr[16],alt="Alt",ymin=0.0,ymax=1.0,xmin=0,xmax=400,colbychrom=TRUE)
  
  for(chrom in altchr){
    pngfile=paste(settings$pngbase,chrom,"png",sep=".")
    multiSNPPlot(pngfile,chrom,alt="Alt")
  }
  
  for(chrom in refchr){
    pngfile=paste(settings$pngbase,chrom,"png",sep=".")
    multiSNPPlot(pngfile,chrom,alt="Ref")
  }
  
}


if(settings$testing){
  

  #### Gal, Xyl and P12/P6 Data !
  galbase = "/Volumes/bioinf/aperezbercoff/Microbiogen-Nov14/dev/2016-06-21.PileupSNPMap/Gal.vs.Glu"
  xylbase = "/Volumes/bioinf/aperezbercoff/Microbiogen-Nov14/dev/2016-06-21.PileupSNPMap/Xyl.vs.Glu"
  popbase = "/Volumes/bioinf/aperezbercoff/Microbiogen-Nov14/dev/2016-06-21.PileupSNPMap/P12.v.P6"
  
  galdata = read.table( paste(galbase,".combinedsnp.tdt",sep=""), header = T, stringsAsFactors = T, sep = '\t', quote = '')
  summary(galdata)
  xyldata = read.table( paste(xylbase,".combinedsnp.tdt",sep=""), header = T, stringsAsFactors = T, sep = '\t', quote = '')
  summary(xyldata)
  popdata = read.table( paste(popbase,".combinedsnp.tdt",sep=""), header = T, stringsAsFactors = T, sep = '\t', quote = '')
  summary(popdata)
  
  
  
  ### Multipanel Plot ##
  
  
  ### ~ Generate multipanel plots for each assembled contig ~~~~~~~~~~~~~ ###
  for(chrom in altchr){
    pngfile=paste(settings$pngbase,chrom,"png",sep=".")
    #CairoPNG(filename=pngfile, width=pwidth, height=pheight, pointsize=12)
    png(filename=pngfile, width=settings$pngwidth, height=settings$pngheight, units = "px", pointsize=settings$pointsize)
    par(mar=c(2,6,2,1))
    #panels = c(1,1,1,2,2,2,3,3,1,1,1,2,2,2,4,4,1,1,1,2,2,2,5,5,1,1,1,2,2,2,6,6)
    #layout(matrix(panels,byrow=FALSE,nrow=8))
    panels = c(1,2,3,4)
    layout(matrix(panels,byrow=FALSE,nrow=4))
    chromSNPFreqPlot(galdata,cnvdata,chrom,alt="Alt",comparison="Gal vs Glu")
    chromSNPFreqPlot(xyldata,cnvdata,chrom,alt="Alt",comparison="Xyl vs Glu")
    chromSNPFreqPlot(popdata,cnvdata,chrom,alt="Alt",comparison="P12 vs P6")
    par(mar=c(5,6,1,1))
    chromCNVPlot(cnvdata,chrom)
    dev.off()
  }
  
  for(chrom in refchr){
    pngfile=paste(settings$pngbase,chrom,"png",sep=".")
    #CairoPNG(filename=pngfile, width=pwidth, height=pheight, pointsize=12)
    png(filename=pngfile, width=settings$pngwidth, height=settings$pngheight, units = "px", pointsize=settings$pointsize)
    par(mar=c(2,6,2,1))
    #panels = c(1,1,1,2,2,2,3,3,1,1,1,2,2,2,4,4,1,1,1,2,2,2,5,5,1,1,1,2,2,2,6,6)
    #layout(matrix(panels,byrow=FALSE,nrow=8))
    panels = c(1,2,3,4)
    layout(matrix(panels,byrow=FALSE,nrow=4))
    chromSNPFreqPlot(galdata,cnvdata,chrom,alt="Ref",comparison="Gal vs Glu")
    chromSNPFreqPlot(xyldata,cnvdata,chrom,alt="Ref",comparison="Xyl vs Glu")
    chromSNPFreqPlot(popdata,cnvdata,chrom,alt="Ref",comparison="P12 vs P6")
    par(mar=c(5,6,1,1))
    chromCNVPlot(cnvdata,chrom)
    dev.off()
  }
  
  
  
  
  chrom = altchr[16]
  
  settings$pcut = 0.001
  multiSNPPlot(paste(settings$pngbase,"exampleXV","png",sep="."),altchr[16],alt="Alt",ymin=0.0,ymax=1.0,xmin=0,xmax=400)
  
  for(chrom in altchr){
    pngfile=paste(settings$pngbase,chrom,"png",sep=".")
    multiSNPPlot(pngfile,chrom,alt="Alt")
  }
  
  for(chrom in refchr){
    pngfile=paste(settings$pngbase,chrom,"png",sep=".")
    multiSNPPlot(pngfile,chrom,alt="Ref")
  }
  
  
  
  
  
  ### >>> Biallelic Run >>>>>
  
  #### Gal, Xyl and P12/P6 Data !
  galbase = "/Users/redwards/Documents/Projects/Microbiogen-Nov14/dev/2016-06-21.PileupSNPMap/Gal.vs.Glu.biallelic"
  xylbase = "/Users/redwards/Documents/Projects/Microbiogen-Nov14/dev/2016-06-21.PileupSNPMap/Xyl.vs.Glu.biallelic"
  popbase = "/Users/redwards/Documents/Projects/Microbiogen-Nov14/dev/2016-06-21.PileupSNPMap/P12.vs.P6.biallelic"
  
  
  
  
  ### >> 11a Alleles Only >>
  
  makeMBGPlots()
  
  makeMBGPlots("biallelic-11a-all")

}


