######################################################
### PacBio Assembly Analysis Image Generator ~~~~~ ###
### VERSION: 0.3.0                           ~~~~~ ###
### LAST EDIT: 17/07/15                      ~~~~~ ###
### AUTHORS: Richard Edwards 2015            ~~~~~ ###
### CONTACT: richard.edwards@unsw.edu.au     ~~~~~ ###
######################################################

############# ::: HISTORY ::: ##################
# v0.2.0 : Added initial summary figures
# v0.3.0 : Updated summary figures to include consistent colouring and chromosome/contig mapping

############# ::: GENERAL SETUP ::: ##################

# Usage = Rscript rje.r pacbio basefile [options] 

setTesting = function(){
  arglist = list("rtype"="pacbio","basefile"="/Users/redwards/Documents/Projects/YeastPacBio-Jun15/analysis/2015-07-14.AssemblyStats/MBG8150.016452.hcq.sgd.ref","refbase"="/Users/redwards/Documents/Projects/YeastPacBio-Jun15/data/2015-07-01.S288C.8150.Ref/sgd.ref")
}

## ~ Read commandline arguments ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ##
rtype = arglist$rtype        # arglist is generated by rje.r
basefile = arglist$basefile  # arglist is generated by rje.r
print(basefile)
refbase = arglist$refbase
print(refbase)

## ~ Set up global parameters and file names ~~~~~~~~~~~~~~~~~~~~~~~~~~ ##
settings = list(basefile=basefile)   # Access with settings[["basefile"]] or settings$basefile
settings$refbase = refbase 
settings$assessment = TRUE # Process outputs from assessment run
settings$covplot = TRUE    # Whether to generate chromosome and contig coverage plots.
settings$pngwidth = 2400
settings$pngheight = 1600
settings$pointsize = 36
settings$minloclen = 1000
settings$topcontigs = 5   # Number of contigs to display as local alignment matches
settings$units = "kb"
settings$plotft = c("gene","mRNA","CDS","rRNA","tRNA","ncRNA","mobile","LTR","origin","centromere","telomere")

# Update default settings from commandline arguments (setting=value)
(arglist)  # This contains settings read in from arguments
arglist = argBool(arglist,c("assessment","covplot"))    # List of Boolean settings
arglist = argNum(arglist,c("pngwidth","pngheight","pointsize","minloclen","topcontigs"))     # List of numeric settings
arglist = argList(arglist,c("plotft"))
for(cmd in names(arglist)){
  settings[[cmd]] = arglist[[cmd]]  # Need to add processing of booleans etc.
}
(settings)  # This contains settings read in from arguments

# Setup base filenames for outputs
settings$basename = strsplit(basefile,'/')[[1]]
(settings$basename = settings$basename[length(settings$basename)])
(settings$pngbase = paste(basefile,".Plots/",settings$basename,sep=""))
dir.create(paste(basefile,".Plots/",sep=""))

# Setup scaling and units
settings$scaling = 1
if(settings$units == "kb"){
  settings$scaling = 1000
}
if(settings$units == "Mb"){
  settings$scaling = 1e6
}

# Setup colour profile
settings$col = list()   # This will have contigs and chromosomes added later.
# Setup feature colour scheme
for(fi in 1:length(settings$plotft)){
  settings$col[[settings$plotft[fi]]] = soton$col[fi+1]
}

(settings)  # This contains final settings

  
############# ::: DEFINE METHODS ::: ##################

## ~ Plot Gridlines ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ##
plotGridlines = function(xmin=0,xmax=0,ymin=0,ymax=0,xminor=0,xmajor=0,yminor=0,ymajor=0,xzero=TRUE,yzero=TRUE,xcap=FALSE,ycap=FALSE){
  # Note that xmin and ymin should only be given if negative
  if(xmin < 0){
    if(xminor > 0){ abline(v=c(1:(xmin/xminor))*xminor,lty="dotted",col="grey") }
    if(xmajor > 0){ abline(v=c(1:(xmin/xmajor))*xmajor,col="grey") }
  }
  if(xmax > 0){
    if(xminor > 0){ abline(v=c(1:(xmax/xminor))*xminor,lty="dotted",col="grey") }
    if(xmajor > 0){ abline(v=c(1:(xmax/xmajor))*xmajor,col="grey") }
  }
  if(xzero == TRUE){ abline(v=0,col="grey") }
  if(xcap == TRUE){ abline(v=xmax,col="grey") }
  if(ymin < 0){
    if(yminor > 0){ abline(h=c(1:(ymin/yminor))*yminor,lty="dotted",col="grey") }
    if(ymajor > 0){ abline(h=c(1:(ymin/ymajor))*ymajor,col="grey") }
  }
  if(ymax > 0){
    if(yminor > 0){ abline(h=c(1:(ymax/yminor))*yminor,lty="dotted",col="grey") }
    if(ymajor > 0){ abline(h=c(1:(ymax/ymajor))*ymajor,col="grey") }
  }
  if(yzero == TRUE){ abline(h=0,col="grey") }
  if(ycap == TRUE){ abline(h=ymax,col="grey") }
}


## ~ Chromosome coverage plot ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ##
chromPlot = function(chromdata,chrom,ctype="Chromosome",minlablen=1000){
  # Plot of the Hits betweeen Chromosomes and Contigs(+) or Chromosomes(-):
  scaling = settings$scaling
  units = paste("(",settings$units,")",sep="")
  plotdata = chromdata[chromdata$Chrom==chrom,]
  plotdata = plotdata[(order(plotdata$Pos)),]
  xmax = max(plotdata$Pos)/scaling
  ymin = -max(plotdata$ChromHit)
  ymax = max(plotdata$HitNum,5)
  # Setup Plot
  plot(c(0,xmax),c(ymin,ymax),col="red",type="n",main=paste(settings$basename,"-",chrom,"-",xmax,settings$units),ylab="BLAST hits",xlab=paste(ctype,"Position",units),xaxs="i")
  plotGridlines(0,xmax,ymin,ymax,10000/scaling,50000/scaling,5,20)
  # Plot Data
  lines(plotdata$Pos/scaling,plotdata$HitNum,col="red",type="l",lwd=2)
  lines(plotdata$Pos/scaling,-plotdata$ChromHit,col="red",type="l",lwd=2)
  lines(plotdata$Pos/scaling,plotdata$ContigNum,col="blue",type="l",lwd=2)
  lines(plotdata$Pos/scaling,-plotdata$ChromNum,col="blue",type="l",lwd=2)
  for(xi in 2:length(plotdata$Pos)){
    xcontig = as.character(plotdata$Contigs[xi])
    if(plotdata$Class[xi-1] == "U" & plotdata$Class[xi] == "U"){
      #rect(plotdata$Pos[xi-1]/scaling,-1,plotdata$Pos[xi]/scaling,1,col=soton$col[3],border="blue",lwd=1)
      rect(plotdata$Pos[xi-1]/scaling,-1,plotdata$Pos[xi]/scaling,1,col=settings$col[[xcontig]],border="blue",lwd=1)
      if((plotdata$Pos[xi] - plotdata$Pos[xi-1]) >= minlablen){
        #text((plotdata$Pos[xi-1]+plotdata$Pos[xi])/(2*scaling),max(2,ymax/20),plotdata$Contigs[xi],pos=4,srt=90,offset=0,col=soton$col[3],xpd=TRUE)
        text((plotdata$Pos[xi-1]+plotdata$Pos[xi])/(2*scaling),max(2,ymax/20),xcontig,pos=4,srt=90,offset=0,col=settings$col[[xcontig]],xpd=TRUE)
      }
    }
    if(plotdata$Class[xi-1] == "C" & plotdata$Class[xi] == "C"){
      rect(plotdata$Pos[xi-1]/scaling,-1,plotdata$Pos[xi]/scaling,1,col=settings$col[[xcontig]],border="blue",lwd=1)
      if((plotdata$Pos[xi] - plotdata$Pos[xi-1]) >= minlablen){
        text((plotdata$Pos[xi-1]+plotdata$Pos[xi])/(2*scaling),max(2,ymax/20),xcontig,pos=4,srt=90,offset=0,col=settings$col[[xcontig]],xpd=TRUE)
      }
    }
    if(plotdata$Class[xi-1] == "N" & plotdata$Class[xi] == "N"){
      points((plotdata$Pos[xi-1]+plotdata$Pos[xi])/(2*scaling),max(1,ymax/10),pch=2,col=soton$col[5],lwd=2)
      text((plotdata$Pos[xi-1]+plotdata$Pos[xi])/(2*scaling),max(2.5,ymax/2),paste((plotdata$Pos[xi]-plotdata$Pos[xi-1])/scaling,settings$units),col=soton$col[5],pos=4,srt=90,xpd=TRUE,offset=0)
    }
  }
  #Legend
  legy = 0.95 * ymin
  legx = xmax/20
  text(0,0.95 * ymax,"Assembly vs Reference",pos=4)
  if(ctype == "Chromosome"){
    text(0,legy,"Reference vs Reference",pos=4)
  }else{
    text(0,legy,"Assembly vs Assembly",pos=4)
  }
  lines(c(8:9)*legx,c(legy,legy),col="red",type="l",lwd=2)
  text(9*legx,legy,"Local",pos=4)
  lines(c(10:11)*legx,c(legy,legy),col="blue",type="l",lwd=2)
  text(11*legx,legy,"Chromosome/Contig",pos=4)
}

## ~ Chromosome coverage difference plot ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ##
chromDifPlot = function(chromdata,chrom){
  # Plot of the Hits betweeen Chromosomes and Contigs(+) or Chromosomes(-):
  scaling = settings$scaling
  units = paste("(",settings$units,")",sep="")
  plotdata = chromdata[chromdata$Chrom==chrom,]
  plotdata = plotdata[(order(plotdata$Pos)),]
  xmax = max(plotdata$Pos)/scaling
  ymin = min(plotdata$HitNum-plotdata$ChromHit)
  ymax = max(plotdata$HitNum-plotdata$ChromHit)
  # Setup Plot
  plot(c(0,xmax),c(ymin,ymax),col="red",type="n",ylab="BLAST Hits",main="Assembly - Reference",xlab=paste("Chromosome Position",units),xaxs="i")
  plotGridlines(0,xmax,ymin,ymax,10000/scaling,50000/scaling,2,10)
  # Plot Data
  lines(plotdata$Pos/scaling,plotdata$HitNum-plotdata$ChromHit,col="red",type="l")
  lines(plotdata$Pos/scaling,plotdata$ContigNum-plotdata$ChromNum,col="blue",type="l")
}

## ~ Features plot ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ##
featurePlot = function(chromdata,features,chrom){
  # Plot of the Features on a Chromosome:
  locus = strsplit(chrom,"[_.]")[[1]][4]
  scaling = settings$scaling
  units = paste("(",settings$units,")",sep="")
  plotdata = chromdata[chromdata$Chrom==chrom,]
  plotdata = plotdata[(order(plotdata$Pos)),]
  xmax = max(plotdata$Pos)/scaling
  plotft = c("gene","mRNA","CDS","rRNA","tRNA","ncRNA","mobile","LTR","origin","centromere","telomere")
  ymax = length(plotft)
  # Setup Plot
  plot(c(0,xmax),c(0,ymax),col="red",type="n",main=paste(chrom,"features"),ylab="",xlab=paste("Chromosome Position",units),xaxs="i",yaxt="n")
  plotGridlines(0,xmax,0,ymax,10000/scaling,50000/scaling,1,ymax)
  # Plot Features
  for(fi in 1:ymax){
    y = ymax-fi
    ftype = plotft[fi]
    text(0,y+0.5,ftype,pos=2,xpd=TRUE,cex=0.8)
    ftdata = features[features$feature == ftype & features$locus == locus,]
    if(dim(ftdata)[1] > 0){
      rect(ftdata$start/scaling,y,ftdata$end/scaling,y+1,col=settings$col[[ftype]],border=NA)  
    }
  }
}


## ~ Contig coverage plot ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ##
# This is now obselete as of v0.2.0.
contigPlot = function(contigdata,contig,minlablen=1000){
  # Plot of the Hits betweeen Chromosomes and Contigs(+) or Chromosomes(-):
  scaling = settings$scaling
  units = paste("(",settings$units,")",sep="")
  plotdata = contigdata[contigdata$Contig==contig,]
  plotdata = plotdata[(order(plotdata$Pos)),]
  xmax = max(plotdata$Pos)/scaling
  ymin = -max(plotdata$RevHit)
  ymax = max(plotdata$FwdHit)
  # Setup Plot
  plot(c(0,xmax),c(ymin,ymax),col="red",type="n",main=paste(settings$basename,contig),ylab="Numbers of BLAST (red) and chromosome (blue) hits",xlab=paste("Contig Position",units),xaxs="i")
  plotGridlines(0,xmax,ymin,ymax,10000/scaling,50000/scaling,5,20)
  # Plot Data
  lines(plotdata$Pos/scaling,plotdata$FwdHit,col="red",type="l")
  lines(plotdata$Pos/scaling,-plotdata$RevHit,col="red",type="l")
  lines(plotdata$Pos/scaling,plotdata$FwdChrom,col="blue",type="l")
  lines(plotdata$Pos/scaling,-plotdata$RevChrom,col="blue",type="l")
  for(xi in 2:length(plotdata$Pos)){
    if(plotdata$Class[xi-1] == "U" & plotdata$Class[xi] == "U" & (plotdata$Pos[xi] - plotdata$Pos[xi-1]) >= minlablen){
      rect(plotdata$Pos[xi-1]/scaling,0,plotdata$Pos[xi]/scaling,plotdata$FwdChrom[xi]-plotdata$RevChrom[xi],col="purple",border="blue",lwd=2)
      text((plotdata$Pos[xi-1]+plotdata$Pos[xi])/(2*scaling),ymax/20,plotdata$Chroms[xi],pos=4,srt=90,offset=0)
    }
  }
}


## ~ Local stack plots ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ##
localChromPlot = function(chromdata,chrom,hitlist){
  # Plot of the Features on a Chromosome:
  scaling = settings$scaling
  units = paste("(",settings$units,")",sep="")
  plotdata = chromdata[chromdata$Chrom==chrom,] 
  plotdata = plotdata[(order(plotdata$Pos)),]
  xmax = max(plotdata$Pos)/scaling
  ymax = length(hitlist)
  # Setup Plot
  plot(c(0,xmax),c(0,ymax),col="red",type="n",main=paste(chrom,"top",settings$topcontigs,"contigs hit stack"),ylab="",xlab=paste("Chromosome Position",units),xaxs="i",yaxt="n")
  plotGridlines(0,xmax,0,ymax,10000/scaling,50000/scaling,1,ymax)
  # Plot Features
  for(fi in 1:ymax){
    y = ymax-fi
    hit = hitlist[fi]
    hitname = strsplit(hit,"_")[[1]][1]
    text(0,y+0.5,hitname,pos=2,xpd=TRUE,cex=0.8)
    ftdata = locdata[locdata$Qry == chrom & locdata$Hit == hit,]
    ftdata = ftdata[order(ftdata$Length),]
    dirndata = ftdata[ftdata$Dirn == "Fwd",]
    if(dim(dirndata)[1] > 0){
      rect(dirndata$QryStart/scaling,y+0.5,dirndata$QryEnd/scaling,y+1,col=settings$col[[hit]])  
    }
    dirndata = ftdata[ftdata$Dirn == "Bwd",]
    if(dim(dirndata)[1] > 0){
      rect(dirndata$QryStart/scaling,y,dirndata$QryEnd/scaling,y+0.5,col=settings$col[[hit]])  
    }
  }
}
localContigPlot = function(contigdata,locdata,contig){
  # Plot of the Features on a Chromosome:
  scaling = settings$scaling
  units = paste("(",settings$units,")",sep="")
  plotdata = contigdata[contigdata$Chrom==contig,] 
  plotdata = plotdata[(order(plotdata$Pos)),]
  xmax = max(plotdata$Pos)/scaling
  ymax = length(levels(locdata$Qry))
  # Setup Plot
  plot(c(0,xmax),c(0,ymax),col="red",type="n",main=paste(contig,"hit stack"),ylab="",xlab=paste("Contig Position",units),xaxs="i",yaxt="n")
  plotGridlines(0,xmax,0,ymax,10000/scaling,50000/scaling,0.5,1)
  # Plot Features
  for(fi in 1:ymax){
    y = ymax-fi
    chrom = levels(locdata$Qry)[fi]
    cname = strsplit(chrom,"_")[[1]][1]
    text(0,y+0.5,cname,pos=2,xpd=TRUE,cex=0.8)
    ftdata = locdata[locdata$Hit == contig & locdata$Qry == chrom,]
    ftdata = ftdata[order(ftdata$Length),]
    dirndata = ftdata[ftdata$Dirn == "Fwd",]
    if(dim(dirndata)[1] > 0){
      rect(dirndata$SbjStart/scaling,y+0.5,dirndata$SbjEnd/scaling,y+1,col=settings$col[[chrom]])  
    }
    dirndata = ftdata[ftdata$Dirn == "Bwd",]
    if(dim(dirndata)[1] > 0){
      rect(dirndata$SbjStart/scaling,y,dirndata$SbjEnd/scaling,y+0.5,col=settings$col[[chrom]])  
    }
  }
}

## ~ DotPlot of local alignments ~~~~~~~~~~~~~~~~~~~~~~~~ ##
# 100% = black
# >= 99% = purple
# >= 90% = blue
# < 90% = red
(pcol = c(rep("red",89),rep("blue",9),"purple","black"))
dotplot = function(locdata,qry,hit,qlen,hlen,pcol=rep(soton$col[1],100),minfrag=0){
  ### ~ Setup Plot ~ ###
  scaling = settings$scaling
  units = paste("(",settings$units,")",sep="")
  #title(main=paste(hit," vs ",qry," (Minfrag=",minfrag,"bp)",sep=""))
  ### ~ Plot local hit data ~ ###
  ldata = locdata[locdata$Qry==qry & locdata$Hit==hit & locdata$Length >= minfrag,]
  ldata$Perc = ldata$Identity / ldata$Length
  qry = strsplit(qry,"_")[[1]][1]
  hit = strsplit(hit,"_")[[1]][1]
  plot(c(1,qlen)/scaling,c(1,hlen)/scaling,type="n",xlab=paste(qry,units),ylab=paste(hit,units),axes=TRUE,ann=TRUE,mar=c(0,1,4,1))
  plotGridlines(0,qlen/scaling,0,hlen/scaling,50000/scaling,200000/scaling,50000/scaling,200000/scaling,xcap=TRUE,ycap=TRUE)
  text(0,0.95*hlen/scaling,paste("minfrag =",minfrag),pos=4)
  for(i in 1:length(ldata$Perc)){
    (tran = as.integer(100*ldata$Perc[i]) - 1)
    #(ldata[i,])
    #icol = (paste(pcol,tran,sep=""))
    icol = pcol[tran+1]
    lines(c(ldata$QryStart[i],ldata$QryEnd[i])/scaling,c(ldata$SbjStart[i],ldata$SbjEnd[i])/scaling,type="l",lwd=2,col=icol)
    lines(c(ldata$QryStart[i],ldata$QryEnd[i])/scaling,c(ldata$SbjStart[i],ldata$SbjEnd[i])/scaling,type="p",lwd=2,col=icol)
  }
}


############# ::: MAIN RUN CODE ::: ##################

## ~ Load and tidy data ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ##

## ~ Overal summary data ~ ##
(summdata = read.table( paste(basefile,".summary.tdt",sep=""), header = T, stringsAsFactors = T, sep = '\t', quote = ''))

## ~ Reference Coverage Summary
rcovdata = read.table( paste(basefile,".Reference.Coverage.tdt",sep=""), header = T, stringsAsFactors = T, sep = '\t', quote = '')
summary(rcovdata)
# Update colour dicionary for chromosomes
cx = 0
for(chrom in rcovdata[order(rcovdata$Length,decreasing=TRUE),"Qry"]){
  cx = cx + 1
  if(cx > length(soton$col)){ cx = 1 }
  settings$col[[chrom]] = soton$col[cx]
  cname = strsplit(chrom,"_")[[1]][1]
  settings$col[[cname]] = soton$col[cx]
}

## ~ Assembly Coverage Summary ~ ##
acovdata = read.table( paste(basefile,".Assembly.Coverage.tdt",sep=""), header = T, stringsAsFactors = T, sep = '\t', quote = '')
summary(acovdata)
# Update colour dicionary for contigs
cx = 0
for(chrom in acovdata[order(acovdata$Length,decreasing=TRUE),"Qry"]){
  cx = cx + 1
  if(cx > length(soton$col)){ cx = 1 }
  settings$col[[chrom]] = soton$col[cx]
  cname = strsplit(chrom,"_")[[1]][1]
  settings$col[[cname]] = soton$col[cx]
}
print(settings$col)

## ~ Reference chromosome coverage ~ ##
chromdata = read.table( paste(basefile,".covplot.chrom.tdt",sep=""), header = T, stringsAsFactors = T, sep = '\t', quote = '')
summary(chromdata)

# ~ Assembly contig coverage ~ ##
contigdata = read.table( paste(basefile,".covplot.contig.tdt",sep=""), header = T, stringsAsFactors = T, sep = '\t', quote = '')
summary(contigdata)

## ~ Load directional contig plot. (Obselete?) ~ ##
dirndata = read.table( paste(basefile,".dirplot.contig.tdt",sep=""), header = T, stringsAsFactors = T, sep = '\t', quote = '')
summary(dirndata)

## ~ Reference vs Assembly local BLAST hits ~ ##
locdata = read.table( paste(basefile,".GABLAM/",settings$basename,".local.tdt",sep=""), header = T, stringsAsFactors = T, sep = '\t', quote = '')
locdata = locdata[locdata$Length >= settings$minloclen,]
locdata$Dirn = "Fwd"
locdata[locdata$SbjStart > locdata$SbjEnd,]$Dirn = "Bwd"
locdata$QryCol = ""
for(chrom in locdata$Qry){
  locdata[locdata$Qry==chrom,][["QryCol"]] = settings$col[[as.character(chrom)]]
}
locdata$HitCol = ""
for(chrom in locdata$Hit){
  locdata[locdata$Hit==chrom,][["HitCol"]] = settings$col[[as.character(chrom)]]
}
summary(locdata)

# ~ Reference genome features ~ ##
features = read.table( paste(refbase,".Feature.tdt",sep=""), header = T, stringsAsFactors = F, sep = '\t', quote = '')
features = features[features$feature != "source",]
features = features[features$feature != "misc_RNA",]
features = features[features$feature != "misc_feature",]
features[features$feature == "mobile_element",]$feature = "mobile"
features[features$feature == "rep_origin",]$feature = "origin"
features$feature = as.factor(features$feature)
summary(features)
(levels(features$feature))


### ~ Generate multipanel plots for each reference chromosome ~~~~~~~~~~~~~ ###
for(chrom in levels(chromdata$Chrom)){
  pngfile=paste(settings$pngbase,"covplot",chrom,"png",sep=".")
  #CairoPNG(filename=pngfile, width=pwidth, height=pheight, pointsize=12)
  png(filename=pngfile, width=settings$pngwidth, height=settings$pngheight, units = "px", pointsize=settings$pointsize)
  panels = c(rep(1,8),rep(2,6),rep(3,4),rep(4,5))
  layout(matrix(panels,byrow=TRUE,nrow=length(panels)))
  par(mar=c(2,6,2,1))
  chromPlot(chromdata,chrom,minlablen=settings$minloclen)
  featurePlot(chromdata,features,chrom)
  covsum = c()
  for(contig in levels(locdata$Hit)){
    #hitlen = max(contigdata[contigdata$Chrom==contig,]$Pos) 
    sdata = locdata[locdata$Qry==chrom & locdata$Hit==contig,]
    #covsum = c(covsum,sum(sdata$Length)/hitlen)
    covsum = c(covsum,sum(sdata$Length))
  }
  (covsum/10000)
  (levels(locdata$Hit))
  (hitlist = levels(locdata$Hit)[order(covsum,decreasing=TRUE)[1:settings$topcontigs]])
  localChromPlot(chromdata,chrom,hitlist)
  par(mar=c(5,6,2,1))
  chromDifPlot(chromdata,chrom)
  dev.off()
}

### ~ Generate multipanel plots for each assembled contig ~~~~~~~~~~~~~ ###
for(contig in levels(contigdata$Chrom)){
  pngfile=paste(settings$pngbase,"covplot",contig,"png",sep=".")
  #CairoPNG(filename=pngfile, width=pwidth, height=pheight, pointsize=12)
  png(filename=pngfile, width=settings$pngwidth, height=settings$pngheight, units = "px", pointsize=settings$pointsize)
  par(mar=c(2,6,2,1))
  #panels = c(1,1,1,2,2,2,3,3,1,1,1,2,2,2,4,4,1,1,1,2,2,2,5,5,1,1,1,2,2,2,6,6)
  #layout(matrix(panels,byrow=FALSE,nrow=8))
  panels = c(1,2,3,1,2,4,1,2,5,1,2,6)
  layout(matrix(panels,byrow=FALSE,nrow=3))
  chromPlot(contigdata,contig,ctype="Contig")
  localContigPlot(contigdata,locdata,contig)
  covsum = c()
  for(chrom in levels(locdata$Qry)){
    sdata = locdata[locdata$Qry==chrom & locdata$Hit==contig,]
    covsum = c(covsum,sum(sdata$Length))
  }
  (hitlist = levels(locdata$Qry)[order(covsum,decreasing=TRUE)[1:4]])
  par(mar=c(5,6,1,1))
  for(chrom in hitlist){  # Have chrom as Query otherwise need to reverse Qry/Hit(Sbj) data - also allows alignments of contig
    dotplot(locdata,chrom,contig,qlen=max(chromdata[chromdata$Chrom==chrom,]$Pos),hlen=max(contigdata[contigdata$Chrom==contig,]$Pos),pcol=pcol,minfrag=settings$minloclen)
  }
  dev.off()
}

# >>> OLD two panel contig plot >>>
#for(contig in levels(contigdata$Chrom)){
#  pngfile=paste(settings$pngbase,"covplot",contig,"png",sep=".")
#  png(filename=pngfile, width=settings$pngwidth, height=settings$pngheight, units = "px", pointsize=settings$pointsize)
#  par(mar=c(2,6,2,1))
#  panels = c(1,1,1,2,2,2)
#  layout(matrix(panels,byrow=TRUE,nrow=6))
#  chromPlot(contigdata,contig,ctype="Contig")
#  localContigPlot(contigdata,locdata,contig)
#  dev.off()
#}
# <<< OLD two panel contig plot <<<

# >>> Alternative directional two panel contig plot >>>
#for(contig in levels(dirndata$Contig)){
#  pngfile=paste(settings$pngbase,"dirnplot",contig,"png",sep=".")
#  png(filename=pngfile, width=settings$pngwidth, height=settings$pngheight, units = "px", pointsize=settings$pointsize)
#  panels = c(1,1,1,1,2,2,2,2)
#  layout(matrix(panels,byrow=TRUE,nrow=8))
#  par(mar=c(2,6,2,1))
#  contigPlot(dirndata,contig)
#  localContigPlot(contigdata,locdata,contig)
#  dev.off()
#}
# <<< Alternative directional two panel contig plot <<<


############# ::: SUMMARY DATA CODE ::: ##################


summGraph = function(summdata){
  scaling = settings$scaling
  # Setup Plot
  xmax = 100 # max(summdata$Length)/scaling - this will be for individual chromosomes
  ymax = dim(summdata)[1]
  par(mar=c(5,6,2,1))
  plot(c(0,xmax),c(0,ymax),col="red",type="n",main=paste(settings$basename,"summary"),ylab="",xlab=paste("Percentage"),xaxs="i",yaxt="n")
  plotGridlines(0,xmax,0,ymax,5,10,0.5,1)
  # Plot Summary
  sumtypes = c("Reference","Assembly","Self","Genes","Genes.Reciprocal","Proteins","Proteins.Reciprocal")
  sumtext = c("Reference","Assembly","Ref.Self","Ref.Genes","Map.Genes","Ref.Proteins","Map.Proteins")
  for(fi in 1:ymax){
    y = ymax-fi
    text(0,y+0.5,sumtext[fi],pos=2,xpd=TRUE,cex=0.8)
    ydata = summdata[summdata$Summary==sumtypes[fi],]
    rect(0,y+0.5,100.0*ydata$Coverage/ydata$Length,y+1,col=soton$col[12])  
    text(0,y+0.75,paste(format(100.0*ydata$Coverage/ydata$Length,digits=5),"% coverage / ",(ydata$Length-ydata$Coverage)/scaling,settings$units," missing",sep=""),pos=4,xpd=TRUE,cex=0.8)
    rect(0,y+0.5,100.0*ydata$Identity/ydata$Length,y,col=soton$col[10])  
    text(0,y+0.25,paste(format(100.0*ydata$Identity/ydata$Length,digits=5),"% identity / ",(ydata$Coverage-ydata$Identity),"bp errors (",ydata$Perfect," of ",ydata$N," perfect)",sep=""),pos=4,xpd=TRUE,cex=0.8)
  }
}

## ~ coverageByGenome mapping of locdb onto all chromosomes, by size ~ ##
coverageByChromosome = function(covdata,locdata,features,gtype="Reference"){
  scaling = settings$scaling
  # Setup Plot
  xmax = max(covdata$Length)/scaling #- this will be for individual chromosomes
  ymax = dim(covdata)[1]
  par(mar=c(5,6,2,1))
  plot(c(0,xmax*1.2),c(0,ymax),col="red",type="n",main=paste(settings$basename,gtype,"coverage"),ylab="",xlab=paste("Length (",settings$units,")",sep=""),xaxs="i",yaxt="n")
  plotGridlines(0,xmax,0,ymax,50,200,0.25,1)
  # Setup list of chromosomes, abbrev csome names and loci for features ("Reference" only)
  chroms = covdata[order(covdata$Length,decreasing=TRUE),]$Qry
  csomes = c()
  clocus = c()
  for(chrom in chroms){
    csomes = c(csomes,strsplit(chrom,"_")[[1]][1])
    clocus = c(clocus,strsplit(chrom,"[_.]")[[1]][4])
  }
  # Plot chromosomes and, features and local mappings
  for(fi in 1:ymax){
    y = ymax-fi
    xlen = covdata[covdata$Qry==chroms[fi],"Length"]/scaling
    chromcol = settings$col[[as.character(chroms[fi])]]
    ydata = covdata[covdata$Qry==chroms[fi],]
    text(0,y+0.5,csomes[fi],pos=2,xpd=TRUE,cex=0.8,col=chromcol)
    text(xlen,y+0.75,paste(format(100.0*ydata$Coverage/ydata$Length,digits=5),"% coverage / ",(ydata$Length-ydata$Coverage)/scaling,settings$units," missing",sep=""),pos=4,xpd=TRUE,cex=1 - 0.015*ymax)
    text(xlen,y+0.25,paste(format(100.0*ydata$Identity/ydata$Length,digits=5),"% identity / ",(ydata$Coverage-ydata$Identity),"bp variant",sep=""),pos=4,xpd=TRUE,cex=1 - 0.015*ymax)
    # Plot chromosome length and (possibly) feature data
    if(gtype == "Reference"){
      ldata = locdata[locdata$Qry == chroms[fi],]
      rect(0,y+0.5,xlen,y+0.75,col="white")
      for(ftype in settings$plotft[settings$plotft != "mRNA"]){
        ftdata = features[features$feature == ftype & features$locus == clocus[fi],]
        if(dim(ftdata)[1] > 0){
          rect(ftdata$start/scaling,y+0.5,ftdata$end/scaling,y+0.75,col=settings$col[[ftype]],border=NA)  
        }
      }      
    }else{
      ldata = locdata[locdata$Hit == chroms[fi],]
      rect(0,y+0.5,xlen,y+0.75,col=chromcol)
    }
    # Plot mapped local hits
    ldata = ldata[order(ldata$Length),]
    dirndata = ldata[ldata$Dirn == "Fwd",]
    if(dim(dirndata)[1] > 0){
      if(gtype == "Reference"){
        rect(dirndata$QryStart/scaling,y+0.75,dirndata$QryEnd/scaling,y+1,col=dirndata$HitCol)  
      }else{
        rect(dirndata$SbjStart/scaling,y+0.75,dirndata$SbjEnd/scaling,y+1,col=dirndata$QryCol)  
      }
    }
    dirndata = ldata[ldata$Dirn == "Bwd",]
    if(dim(dirndata)[1] > 0){
      if(gtype == "Reference"){
        rect(dirndata$QryStart/scaling,y+0.25,dirndata$QryEnd/scaling,y+0.5,col=dirndata$HitCol)  
      }else{
        rect(dirndata$SbjStart/scaling,y+0.25,dirndata$SbjEnd/scaling,y+0.5,col=dirndata$QryCol)  
      }
    }
  }
}


coverageByChromosome(rcovdata,locdata,features,gtype="Reference")
coverageByChromosome(acovdata,locdata,features,gtype="Assembly")

## ~ SUMMARY DATA ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ##

## ~ Two-panel plot of overal summary and chromosome coverage and contig coverage ~ ##
pngfile=paste(settings$pngbase,"summary","png",sep=".")
#CairoPNG(filename=pngfile, width=pwidth, height=pheight, pointsize=12)
png(filename=pngfile, width=settings$pngwidth, height=settings$pngheight, units = "px", pointsize=settings$pointsize)
panels = c(rep(1,3),rep(2,5))
layout(matrix(panels,byrow=TRUE,nrow=length(panels)))
par(mar=c(5,6,1,1))
summGraph(summdata)
coverageByChromosome(rcovdata,locdata,features,gtype="Reference")
dev.off()
## ~ Plot contig coverage ~ ##
pngfile=paste(settings$pngbase,"assembly","png",sep=".")
#CairoPNG(filename=pngfile, width=pwidth, height=pheight, pointsize=12)
png(filename=pngfile, width=settings$pngwidth, height=settings$pngheight, units = "px", pointsize=settings$pointsize)
par(mar=c(5,6,1,1))
coverageByChromosome(acovdata,locdata,features,gtype="Assembly")
dev.off()
