########################################################
### RJE Sequence R Scripts                     ~~~~~ ###
### VERSION: 0.1.0                             ~~~~~ ###
### LAST EDIT: 13/11/15                        ~~~~~ ###
### AUTHORS: Richard Edwards 2015              ~~~~~ ###
### CONTACT: richard.edwards@unsw.edu.au       ~~~~~ ###
########################################################

################# ::: HISTORY ::: ######################
# v0.1.0 : Initial version for simple GC plot.

############### ::: GENERAL SETUP ::: ##################
# Usage = Rscript rje.r seq basefile [options] 

setTesting = function(){
  arglist = list("rtype"="seq","basefile"="/Users/redwards/ownCloud/projects/ManefieldPacBio-Sep15/analysis/2015-09-14.WON710A1.SP16514.hcq0/WON710A1.SP16514.hcq0")
  arglist = list("rtype"="seq","basefile"="/Users/redwards/Data/projects/ManefieldPacBio-Sep15/analysis/2015-11-05.Unitig0/WON710A1.SP16546.hcq0")
}

## ~ Read commandline arguments ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ##
rtype = arglist$rtype        # arglist is generated by rje.r
basefile = arglist$basefile  # arglist is generated by rje.r
print(basefile)

## ~ Set up global parameters and file names ~~~~~~~~~~~~~~~~~~~~~~~~~~ ##
settings = list(basefile=basefile)   # Access with settings[["basefile"]] or settings$basefile
settings$pngwidth = 2400
settings$pngheight = 1600
settings$pointsize = 36

# Update default settings from commandline arguments (setting=value)
(arglist)  # This contains settings read in from arguments
arglist = argBool(arglist,c())    # List of Boolean settings
arglist = argNum(arglist,c("pngwidth","pngheight","pointsize"))     # List of numeric settings
arglist = argList(arglist,c())
for(cmd in names(arglist)){
  settings[[cmd]] = arglist[[cmd]]  # Need to add processing of booleans etc.
}
(settings)  # This contains settings read in from arguments

# Setup colour profile
# Setup feature colour scheme
#settings$col = list()   # This will have contigs and chromosomes added later.
#  settings$col[[settings$plotft[fi]]] = soton$col[fi+1]
#for(fi in 1:length(settings$plotft)){
#}

#(settings)  # This contains final settings

############# ::: LOAD AND TIDY DATA ::: ##################
#? Should this be made a function?

## ~ Genome NT window freq data ~ ##
ReadGCData = function(gcfile){
  gcdata = read.table( gcfile, header = T, stringsAsFactors = T, sep = '\t', quote = '')
  summary(gcdata)
  gcdata$GC = gcdata$G + gcdata$C
  gcdata$GSkew = (gcdata$G - gcdata$C) / gcdata$GC
  gcdata$AT = gcdata$A + gcdata$T
  gcdata$ASkew = (gcdata$A - gcdata$T) / gcdata$AT
  return(gcdata)
}
  
GSkewPlot = function(gcdata,xrange=c(0,-1)){
  meangc = mean(gcdata$GC)
  xmin = xrange[1]
  xmax = max(gcdata$Pos)
  if(xrange[2] > 0){
    xmax = xrange[2]
  }
  ymax = max(c(gcdata$GSkew,gcdata$ASkew))
  ymin = min(c(gcdata$GSkew,gcdata$ASkew))
  plot(gcdata$Pos,gcdata$GC-meangc,type="l",ylim=c(ymin,ymax),xlim=c(xmin,xmax),xlab="Genome Position",ylab="Skew")
  abline(h=0.0,lty="dotted",col="grey")
  points(gcdata$Pos,gcdata$GSkew,type="l",col="red")
  points(gcdata$Pos,gcdata$ASkew,type="l",col="blue")
  text(xmin,ymax,"GC Bias")
  text(xmin,ymax*0.9,"G-C Skew",col="red")
  text(xmin,ymax*0.8,"A-T Skew",col="blue")
}



pngfile=paste(settings$basefile,"gc_plot","png",sep=".")
png(filename=pngfile, width=settings$pngwidth, height=settings$pngheight, units = "px", pointsize=settings$pointsize)
par(mar=c(5,6,1,1))
panels = 1:3
layout(matrix(panels,byrow=TRUE,nrow=3))

## ~ Genome NT window freq data ~ ##
gcdata = ReadGCData( paste(basefile,".1e6.genome_nt.tdt",sep=""))
GSkewPlot(gcdata)
gcdata = ReadGCData( paste(basefile,".1e5.genome_nt.tdt",sep=""))
GSkewPlot(gcdata)
gcdata = ReadGCData( paste(basefile,".1e4.genome_nt.tdt",sep=""))
GSkewPlot(gcdata)

dev.off()




basefile = "/Users/redwards/Data/projects/ManefieldPacBio-Sep15/analysis/2015-11-05.Unitig0/WON710A1.SP16546.hcq0.Mb2-3"

pngfile=paste(settings$basefile,"gc_plot","png",sep=".")
png(filename=pngfile, width=settings$pngwidth, height=settings$pngheight, units = "px", pointsize=settings$pointsize)
panels = 1:3
layout(matrix(panels,byrow=TRUE,nrow=3))
xrange=c(2.4e6,2.6e6)
for(ntx in c("1e5","1e4","1e3")){
  gcdata = ReadGCData( paste(basefile,".",ntx,".genome_nt.tdt",sep=""))
  GSkewPlot(gcdata,xrange)
}
dev.off()
  
  
  
  
  
  
  
## ~ Genome NT window freq data ~ ##
gcdata = read.table( paste(basefile,".1e6.genome_nt.tdt",sep=""), header = T, stringsAsFactors = T, sep = '\t', quote = '')
summary(gcdata)
gcdata$GC = gcdata$G + gcdata$C
gcdata$AT = gcdata$A + gcdata$T

meangc = mean(gcdata$GC)
plot(gcdata$Pos,gcdata$GC,type="l")
abline(h=meangc)

## ~ Genome NT window freq data ~ ##
gcdata = read.table( paste(basefile,".1e5.genome_nt.tdt",sep=""), header = T, stringsAsFactors = T, sep = '\t', quote = '')
summary(gcdata)
gcdata$GC = gcdata$G + gcdata$C
gcdata$AT = gcdata$A + gcdata$T

meangc = mean(gcdata$GC)
plot(gcdata$Pos,gcdata$GC,type="l")
abline(h=meangc)

## ~ Genome NT window freq data ~ ##
gcdata = read.table( paste(basefile,".1e4.genome_nt.tdt",sep=""), header = T, stringsAsFactors = T, sep = '\t', quote = '')
summary(gcdata)
gcdata$GC = gcdata$G + gcdata$C
gcdata$AT = gcdata$A + gcdata$T

meangc = mean(gcdata$GC)
plot(gcdata$Pos,gcdata$GC,type="l")
abline(h=meangc)

dev.off()
