###################################################
### SNP Frequency Plotting Tool - R Graphics ~~ ###
### VERSION: 0.5.0                        ~~~~~ ###
### LAST EDIT: 24/11/17                   ~~~~~ ###
### AUTHORS: Richard Edwards 2017         ~~~~~ ###
### CONTACT: richard.edwards@unsw.edu.au  ~~~~~ ###
###################################################

################# ::: HISTORY ::: ######################
# v0.0.0 : Initial version based on PAGSAT.R v0.7.3.
# v0.1.0 : Modifying to run from commandline. Deleted excess PAGSAT code. Renamed from Snapper.R.
# v0.2.0 : Added fdrplot=T/F as an option. Fixed CNV plot zoom bug.
# v0.3.0 : Added multiplot=T/F as an option.
# v0.4.0 : Added feature plots for reference and fixed FDR plotting.
# v0.5.0 : Updated for plots lacking Snapper data.


#!# Rearrange a little and make more elements into functions (load data etc.)
#!# Move functions shared with PAGSAT into a separate file
#!# Add loading of regions to mark from separate table
#!# Plot 0CNV regions on the SNPFreq plots

############### ::: GENERAL SETUP ::: ##################
# Usage = Rscript rje.r snpfreq basefile [options] 
# Main options: snapbase=X    : base path to Snapper files
#               freqpath=X    : path to Pileup frequency outputs
#               freqcomp=LIST : list of comparisons to make
#               chrom=X, xmin=X, xmax=X : Limit plots to a specific chromosome and/or region
#               pcut=X        : Minimum p-value to shade in yellow [0.01]

#i# As this is the Snapper.R script, basefile should refer to the base of the Snapper run.
#i# |-- V0.1.0 uses the *.cnv.tdt table only. (This might be for a custom plot.)
#i# NOTE: This is a change from V0.0.0, in which this was assigned to snpbase and basefile was used for the SNP frequency runs

#># SNP Frequency run data is now set by freqpath=X + freqcomp=LIST
#># |-- *.fdr.tdt
#># |-- *.combinedsnp.tdt

#i# NOTE: Multipanel SNP Frequency plots will need to take multiple freqcomp inputs: use the number to define the number of panels

#># Output file names will be based on a outbase=X, which is now given by basefile=X
#># Input for Snapper files (Snapper basefile) is now given by snapbase=X

setTesting = function(){
  arglist = list("rtype"="snpfreq","rdir"="/Users/redwards/Dropbox/_Repository_/slimsuite/libraries/r/")
  arglist$snapbase = "/Volumes/bioinf-1/aperezbercoff/Microbiogen-Nov14/dev/2016-06-21.PileupSNPMap/MBG11A.MBG16533.vs.sgd.ref"
  arglist$freqpath = "/Volumes/bioinf-1/aperezbercoff/Microbiogen-Nov14/dev/2016-06-21.PileupSNPMap/"
  arglist$freqcomp = "Gal.vs.Glu.biallelic,Xyl.vs.Glu.biallelic,P12.vs.P6.biallelic"
  arglist$basefile = "Gal.vs.Glu.test"
  arglist$refbase = "/Users/redwards/ownCloud/projects/MBGHybridSelection-Feb17/data/2017-03-10.Ref.S288C/sgd.R64.2.1"
  setwd("/Users/redwards/ownCloud/projects/Microbiogen-Nov14/dev/2017-03-20.SNPFreqTest")
}

## ~ Read commandline arguments ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ##
rtype = arglist$rtype        # arglist is generated by rje.r
basefile = arglist$basefile  # arglist is generated by rje.r
#print(basefile)

## ~ Set up global parameters and file names ~~~~~~~~~~~~~~~~~~~~~~~~~~ ##
settings = list(basefile=basefile)  # IO Basefile Access with settings[["basefile"]] or settings$basefile
settings$units = "kb"
settings$pngwidth = 2400
settings$pngheight = 1600
settings$dualpng = 1600
settings$pointsize = 36
settings$pcut = 0.01
settings$fdrplot = FALSE
settings$multiplot = FALSE

## ~ Setup Chromosome region settings ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ##
settings$chrom = NA
settings$ymin=0
settings$ymax=1.0
settings$xmin=0
settings$xmax=-1

## ~ Setup chromosome plot and features methods ~~~~~~~~~~~~~~~~~~~~~~~~ ##
rjesource("rje_genomics.r")   
settings = featureDefaults(settings)

## ~ Update default settings from commandline arguments (setting=value) ~ ##
(arglist)  # This contains settings read in from arguments
arglist = argBool(arglist,c("assessment","covplot","diploid","genesummary","protsummary","chromalign","fdrplot","multiplot"))    # List of Boolean settings
arglist = argNum(arglist,c("pngwidth","pngheight","pointsize","minloclen","topcontigs","pcut","ymin","ymax","xmin","xmax"))     # List of numeric settings
arglist = argList(arglist,c("plotft","freqcomp"))
for(cmd in names(arglist)){
  settings[[cmd]] = arglist[[cmd]]  # Need to add processing of booleans etc.
}
(settings)  # This contains settings read in from arguments

# Setup base filenames for outputs
settings$outbase = settings$basefile
(settings$basename = settings$outbase)          # Output basefile without path
(settings$plotdir = paste(settings$outbase,".Plots/",sep=""))         # Directory for Plot output
(settings$pngbase = paste(settings$plotdir,settings$basename,sep=""))     # Directory for Plot output
if(file.exists(settings$plotdir)==FALSE){
  dir.create(settings$plotdir)
}

# Setup scaling and units
settings$scaling = 1
if(settings$units == "kb"){
  settings$scaling = 1000
}
if(settings$units == "Mb"){
  settings$scaling = 1e6
}

# Setup colour profile
if(settings$features){
  writeLines("Features for plotting:")
  writeLines(settings$plotft)
}
settings = featureColours(settings)
settings$testing=FALSE

#print(settings)  # This contains final settings


############# ::: LOAD AND TIDY DATA ::: ##################

## Load the SNP frequency data for each input comparison
fdrdb = list()  # Each input freqcomp will have an entry
combdb = list()
for(freqcomp in settings$freqcomp){
  fdrdb[[freqcomp]] = read.table( paste(settings$freqpath,freqcomp,".fdr.tdt",sep=""), header = T, stringsAsFactors = T, sep = '\t', quote = '')
  summary(fdrdb[[freqcomp]])
  if(! 'AltPos' %in% colnames(fdrdb[[freqcomp]])){
      fdrdb[[freqcomp]]$AltPos = fdrdb[[freqcomp]]$Pos
      fdrdb[[freqcomp]]$AltLocus = fdrdb[[freqcomp]]$Locus
  }

  if(file.exists(paste(settings$freqpath,freqcomp,".combinedsnp.tdt",sep=""))){
      freqfile = paste(settings$freqpath,freqcomp,".combinedsnp.tdt",sep="")
  }else{
      freqfile = paste(settings$freqpath,freqcomp,".snpcomb.tdt",sep="")
  }
  combdb[[freqcomp]] = read.table( freqfile, header = T, stringsAsFactors = T, sep = '\t', quote = '')
  summary(combdb[[freqcomp]])
  if(! 'AltPos' %in% colnames(combdb[[freqcomp]])){
      combdb[[freqcomp]]$AltPos = combdb[[freqcomp]]$Pos
      combdb[[freqcomp]]$AltLocus = combdb[[freqcomp]]$Locus
  }
}

## Load the Snapper CNV table
snapdb = list()  # List of each Snapper output
cnvfile = paste(settings$snapbase,"cnv","tdt",sep=".")
if(file.exists(cnvfile)){
    #i# This is proper Snapper output
    for(ext in c("cnv")){
        snapdb[[ext]] = read.table( paste(settings$snapbase,ext,"tdt",sep="."), header = T, stringsAsFactors = T, sep = '\t', quote = '')
    }
    # Update CNV data and set colours
    cnvdata = snapdb[["cnv"]]
    cnvdata$Chrom = cnvdata$Locus
    writeLines(str(cnvdata))
    writeLines(summary(cnvdata))
    
    levels(cnvdata$Locus)
    altchr = levels(as.factor(as.character(cnvdata[cnvdata$Source=="Alt",]$Locus)))
}else{
    writeLines(paste(settings$snapbase,"Feature.tdt",sep="."))
    cnvdata = read.table( paste(settings$snapbase,"Feature.tdt",sep="."), header = T, stringsAsFactors = T, sep = '\t', quote = '')
    colnames(cnvdata) = c("Locus","feature","position","Start","End","locus_tag","protein_id","details")
    cnvdata$Chrom = cnvdata$Locus
    cnvdata$Source = "Ref"
    writeLines(summary(cnvdata))
    
    levels(cnvdata$Locus)
    altchr = levels(as.factor(as.character(cnvdata[cnvdata$Source=="Ref",]$Locus)))
}



# Update colour dicionary for contigs
cx = 0
for(chrom in altchr){
  cx = cx + 1
  if(cx > length(soton$col)){ cx = 1 }
  settings$col[[chrom]] = soton$col[cx]
  cname = strsplit(chrom,"_")[[1]][1]
  settings$col[[cname]] = soton$col[cx]
}
refchr = levels(as.factor(as.character(cnvdata[cnvdata$Source=="Ref",]$Locus)))
# Update colour dicionary for chromosomes
cx = 0
for(chrom in refchr){
  cx = cx + 1
  if(cx > length(soton$col)){ cx = 1 }
  settings$col[[chrom]] = soton$col[cx]
  cname = strsplit(chrom,"_")[[1]][1]
  settings$col[[cname]] = soton$col[cx]
}

## ~ Load features data ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ##
if(is.null(settings$refbase)){
  if(settings$multiplot == FALSE){ writeLines("No RefBase set - no feature plots") }
  ftdb = list("Ref"=data.frame())
}else{
  ftdata = featureSetup(settings$refbase,ftype="Ref")
  print(table(ftdata$feature))
  ftdb = list("Ref"=ftdata)
}

#i#featurePlot("Reference","BK006934",xmin=1,xmax=300,acconly=FALSE)
#i# NOTE: featurePlot() uses pre-scaled xmin and xmax. 

############# ::: DEFINE METHODS ::: ##################

## ~ Chromosome copy number plot ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ##
chromCNVPlot = function(chromdata,chrom,xmin=0,xmax=-1){
  # Plot of the Hits betweeen Chromosomes and Contigs(+) or Chromosomes(-):
  scaling = settings$scaling
  units = paste("(",settings$units,")",sep="")
  (plotdata = chromdata[chromdata$Chrom==chrom,])
  plotdata = plotdata[(order(plotdata$Start)),]
  if(xmax < 0){
    xmax = max(plotdata$End)/scaling
  }
  ymin = 0
  ymax = max(plotdata$Copy)
  # Setup Plot
  plot(c(xmin,xmax),c(ymin,ymax),col="red",type="n",ylab="Copy Number",main=paste(chrom,"CNV"),xlab=paste("Chromosome Position",units),xaxs="i")
  plotGridlines(xmin,xmax,ymin,ymax,10000/scaling,50000/scaling,1,10)
  # Plot Data
  for(xi in 1:length(plotdata$Start)){
    recdata = plotdata[xi,]
    rect(recdata$Start/scaling,0.0,recdata$End/scaling,recdata$Copy,col=settings$col[[chrom]])
  }#lines(plotdata$Pos/scaling,plotdata$ContigNum-plotdata$ChromNum,col="blue",type="l")
}
# Set test data
if(settings$testing){
  chrom = altchr[3]
  chromdata = cnvdata
  chromCNVPlot(chromdata,chrom)
  chromCNVPlot(chromdata,refchr[3])
}

## ~ Plot changes in SNP frequencies across chromosome ~~~~~~~~~~~~~~~~ ##
#># colbychrom=TRUE uses the colour of the matching chromosome for the points and lines
#># colbychrom=FALSE uses the red for up and blue for down. (Purple for no change.)
chromSNPFreqPlot = function(snpdata,chromdata,chrom,alt="Alt",comparison=settings$basename,ymin=0,ymax=1.0,xmin=0,xmax=-1,colbychrom=TRUE,region=FALSE){
  # Plot of the Hits betweeen Chromosomes and Contigs(+) or Chromosomes(-):
  scaling = settings$scaling
  units = paste("(",settings$units,")",sep="")
  if(alt == "Alt"){
    (plotdata = snpdata[snpdata$AltLocus==chrom,])
    plotdata = plotdata[(order(plotdata$AltPos)),]
  }else{
    (plotdata = snpdata[snpdata$Locus==chrom,])
    plotdata = plotdata[(order(plotdata$Pos)),]
  }
  if(xmax < xmin){
    xmax = max(chromdata[chromdata$Chrom==chrom,]$End)/scaling
  }
  # Setup Plot
  if(region == FALSE){ region = chrom }
  plot(c(xmin,xmax),c(ymin,ymax),col="red",type="n",ylab="SNP Freq",main=paste(region,comparison),xlab=paste("Chromosome Position",units),xaxs="i")
  plotGridlines(xmin,xmax,ymin,ymax,10000/scaling,50000/scaling,0.1,0.5)
  if(dim(plotdata)[1] < 1){ return(0) }
  # Plot Data
  for(xi in 1:length(plotdata$Pos)){
    pdata = plotdata[xi,]
    # Set plot symbol
    if(pdata$MajDiff > 0){
      ppch=24
    }else if(pdata$MajDiff == 0){
      ppch=23
    }else{
      ppch=25
    }
    # Set plot colours
    if(pdata$MajProb <= settings$pcut){
      p = pdata$MajProb
      i = max(0,(10 + log10(p))/10.0)
      pbg = rgb(1,1,i)
      plwd = 2 - i
    }else{
      pbg = NA
      plwd = 1
    }
    if(colbychrom){
      if(alt == "Alt"){
        pchrom = as.character(pdata$Locus)
        pcol = settings$col[[pchrom]]
      }else{
        pchrom = as.character(pdata$AltLocus)
        pcol = settings$col[[pchrom]]
      }
    }else{
      if(pdata$MajDiff > 0){
        i = min(1,max(0.2,2*pdata$MajDiff))
        pcol = rgb(1,1-i,1-i)
      }else if(pdata$MajDiff == 0){
        pcol = rgb(1,0.5,1)
      }else{
        i = max(-1.0,min(-0.2,2*pdata$MajDiff))
        pcol = rgb(1+i,1+i,1)
      }
    }
    # Plot lines and symbols
    if(alt == "Alt"){
      lines(c(pdata$AltPos/scaling,pdata$AltPos/scaling),c(pdata$MajFreq,pdata$MajFreq-pdata$MajDiff),col=pcol,type="l",lwd=plwd)
      points(c(pdata$AltPos/scaling),c(pdata$MajFreq),col=pcol,pch=ppch,bg=pbg)
    }else{
      lines(c(pdata$Pos/scaling,pdata$Pos/scaling),c(pdata$MajFreq,pdata$MajFreq-pdata$MajDiff),col=pcol,type="l",lwd=plwd)
      points(c(pdata$Pos/scaling),c(pdata$MajFreq),col=pcol,pch=ppch,bg=pbg)
    }
  }#lines(plotdata$Pos/scaling,plotdata$ContigNum-plotdata$ChromNum,col="blue",type="l")
}

## ~ Plot changes in SNP frequencies across chromosome, assuming 50% starting frequency ~~~~~~~~~~~~~~~~ ##
#!# Not currently used for anything but can be useful for explanatory plots/talks.
chromSNPFreqExpPlot = function(snpdata,chromdata,chrom,alt="Alt",comparison=settings$basename,ymin=0,ymax=1.0,xmin=0,xmax=-1,colbychrom=TRUE,region=FALSE,inverse=FALSE){
  # Plot of the Hits betweeen Chromosomes and Contigs(+) or Chromosomes(-):
  scaling = settings$scaling
  units = paste("(",settings$units,")",sep="")
  if(alt == "Alt"){
    (plotdata = snpdata[snpdata$AltLocus==chrom,])
    plotdata = plotdata[(order(plotdata$AltPos)),]
  }else{
    (plotdata = snpdata[snpdata$Locus==chrom,])
    plotdata = plotdata[(order(plotdata$Pos)),]
  }
  if(inverse){
    plotdata$MajFreq = plotdata$MajFreq - plotdata$MajDiff
  }
  plotdata$MajDiff = plotdata$MajFreq - 0.5
  plotdata$MajProb = 1.0
  if(xmax < xmin){
    xmax = max(chromdata[chromdata$Chrom==chrom,]$End)/scaling
  }
  # Setup Plot
  if(region == FALSE){ region = chrom }
  plot(c(xmin,xmax),c(ymin,ymax),col="red",type="n",ylab="SNP Freq",main=paste(region,comparison),xlab=paste("Chromosome Position",units),xaxs="i")
  plotGridlines(xmin,xmax,ymin,ymax,10000/scaling,50000/scaling,0.1,0.5)
  if(dim(plotdata)[1] < 1){ return(0) }
  # Plot Data
  for(xi in 1:length(plotdata$Pos)){
    pdata = plotdata[xi,]
    # Set plot symbol
    if(pdata$MajDiff > 0){
      ppch=24
    }else if(pdata$MajDiff == 0){
      ppch=23
    }else{
      ppch=25
    }
    # Set plot colours
    if(pdata$MajProb <= settings$pcut){
      p = pdata$MajProb
      i = max(0,(10 + log10(p))/10.0)
      pbg = rgb(1,1,i)
      plwd = 2 - i
    }else{
      pbg = NA
      plwd = 1
    }
    if(colbychrom){
      if(alt == "Alt"){
        pchrom = as.character(pdata$Locus)
        pcol = settings$col[[pchrom]]
      }else{
        pchrom = as.character(pdata$AltLocus)
        pcol = settings$col[[pchrom]]
      }
    }else{
      if(pdata$MajDiff > 0){
        i = min(1,max(0.2,2*pdata$MajDiff))
        pcol = rgb(1,1-i,1-i)
      }else{
        i = max(-1.0,min(-0.2,2*pdata$MajDiff))
        pcol = rgb(1+i,1+i,1)
      }
    }
    # Plot lines and symbols
    if(alt == "Alt"){
      lines(c(pdata$AltPos/scaling,pdata$AltPos/scaling),c(pdata$MajFreq,pdata$MajFreq-pdata$MajDiff),col=pcol,type="l",lwd=plwd)
      points(c(pdata$AltPos/scaling),c(pdata$MajFreq),col=pcol,pch=ppch,bg=pbg)
    }else{
      lines(c(pdata$Pos/scaling,pdata$Pos/scaling),c(pdata$MajFreq,pdata$MajFreq-pdata$MajDiff),col=pcol,type="l",lwd=plwd)
      points(c(pdata$Pos/scaling),c(pdata$MajFreq),col=pcol,pch=ppch,bg=pbg)
    }
  }#lines(plotdata$Pos/scaling,plotdata$ContigNum-plotdata$ChromNum,col="blue",type="l")
}

# Set test data
if(settings$testing){
  chrom = altchr[3]
  chromdata = cnvdata
  alt="Alt"
  (comparison = settings$freqcomp[1])
  snpdata = fdrdb[[comparison]]
  snpdata = combdb[[comparison]]
  chromSNPFreqPlot(snpdata,chromdata,chrom,alt="Alt",comparison=comparison,ymin=0,ymax=1.0,xmin=0,xmax=-1)
  chromSNPFreqPlot(snpdata,chromdata,chrom,alt="Alt",comparison=comparison,ymin=0,ymax=1.0,xmin=0,xmax=-1,colbychrom = FALSE)
  chromSNPFreqExpPlot(snpdata,chromdata,chrom,alt="Alt",comparison=comparison,ymin=0,ymax=1.0,xmin=0,xmax=-1)
  chromSNPFreqExpPlot(snpdata,chromdata,chrom,alt="Alt",comparison=comparison,ymin=0,ymax=1.0,xmin=0,xmax=-1,colbychrom = FALSE)
}


## ~ Dual panel plot of SNP Frequency and CNV ~~~~~~~~~~~~~~~~ ##
multiSNPCNVPlot = function(pngfile,chrom,alt="Alt",ymin=0,ymax=1.0,xmin=0,xmax=-1,makepng=TRUE){
  if(makepng){
    png(filename=pngfile, width=settings$pngwidth, height=settings$dualpng, units = "px", pointsize=settings$pointsize)
  }
  par(mar=c(2,6,2,1))
  cn = length(settings$freqcomp) + 1
  panels = 1:cn
  layout(matrix(panels,byrow=FALSE,nrow=cn))
  for(ci in 1:(cn-1)){
    if(settings$fdrplot){
      chromSNPFreqPlot(fdrdb[[settings$freqcomp[ci]]],cnvdata,chrom,alt,paste(settings$freqcomp[ci],"FDR"),ymin,ymax,xmin,xmax,colbychrom = TRUE)
    }else{
      chromSNPFreqPlot(combdb[[settings$freqcomp[ci]]],cnvdata,chrom,alt,settings$freqcomp[ci],ymin,ymax,xmin,xmax,colbychrom = TRUE)
    }
  }
  par(mar=c(5,6,1.5,1))
  chromCNVPlot(cnvdata,chrom,xmin=xmin,xmax=xmax) 
  if(makepng){
    dev.off()
  }
}
if(settings$testing){
  chrom = refchr[3]
  multiSNPCNVPlot("pngfile",chrom,alt="Ref",ymin=0,ymax=1.0,xmin=0,xmax=-1,makepng=FALSE)
  chrom = altchr[3]
  multiSNPCNVPlot("pngfile",chrom,alt="Alt",ymin=0,ymax=1.0,xmin=0,xmax=-1,makepng=FALSE)
}

## ~ Mulitpanel plot of different SNP Frequency changes ~~~~~~~~~~~~~~~~ ##
multiSNPPlot = function(pngfile,chrom,alt="Alt",ymin=0,ymax=1.0,xmin=0,xmax=-1,makepng=TRUE){
  # Setup PNG
  if(makepng){
    png(filename=pngfile, width=settings$pngwidth, height=settings$pngheight, units = "px", pointsize=settings$pointsize)
  }
  # Setup Multipanel plot
  par(mar=c(2,6,2,1))
  cn = length(settings$freqcomp)
  cx = cn 
  plotft = FALSE
  if(isFeatureData(alt)){ 
    cx = cn + 1 
    plotft = TRUE
  }
  panels = 1:cx
  layout(matrix(panels,byrow=FALSE,nrow=cx))
  for(ci in 1:cn){
    if(ci == cn && plotft == FALSE){
      par(mar=c(5,6,1.5,1))
    }
    if(settings$fdrplot){
      chromSNPFreqPlot(fdrdb[[settings$freqcomp[ci]]],cnvdata,chrom,alt,paste(settings$freqcomp[ci],"FDR"),ymin,ymax,xmin,xmax,colbychrom = FALSE)
    }else{
      chromSNPFreqPlot(combdb[[settings$freqcomp[ci]]],cnvdata,chrom,alt,settings$freqcomp[ci],ymin,ymax,xmin,xmax,colbychrom = FALSE)
    }
  }
  if(plotft){
    par(mar=c(5,6,1.5,1))
    if(xmax < xmin){
      xmax = max(cnvdata[cnvdata$Chrom==chrom,]$End)/settings$scaling
    }
    featurePlot(alt,chrom,xmin,xmax)
  }  
  if(makepng){
    dev.off()
  }
}
if(settings$testing){
  settings$pcut = 0.001  # Controls which points are shaded yellow
  settings$fdrplot = TRUE
  multiSNPPlot("pngfile",chrom,alt="Ref",ymin=0,ymax=1.0,xmin=0,xmax=-1,makepng=FALSE)
}


### ~ Make functions for the different main plot types

#1# Region Zoom plot of a single chromosome
zoomPlot = function(makepng=TRUE){
  chrom = settings$chrom
  ext = "png"
  if(settings$fdrplot){ ext = "fdr.png" }
  writeLines(chrom)
  if(settings$chrom %in% refchr){
    pngfile=paste(settings$pngbase,chrom,ext,sep=".")
    multiSNPPlot(pngfile,chrom,alt="Ref",ymin=settings$ymin,ymax=settings$ymax,xmin=settings$xmin,xmax=settings$xmax,makepng=makepng)
    pngfile=paste(settings$pngbase,chrom,"cnv",ext,sep=".")
    multiSNPCNVPlot(pngfile,chrom,alt="Ref",ymin=settings$ymin,ymax=settings$ymax,xmin=settings$xmin,xmax=settings$xmax,makepng=makepng)
  }
  if(settings$chrom %in% altchr){
    pngfile=paste(settings$pngbase,chrom,ext,sep=".")
    multiSNPPlot(pngfile,chrom,alt="Alt",ymin=settings$ymin,ymax=settings$ymax,xmin=settings$xmin,xmax=settings$xmax,makepng=makepng)
    pngfile=paste(settings$pngbase,chrom,"cnv",ext,sep=".")
    multiSNPCNVPlot(pngfile,chrom,alt="Alt",ymin=settings$ymin,ymax=settings$ymax,xmin=settings$xmin,xmax=settings$xmax,makepng=makepng)
  }
}  

#2# Standard Solo plots per chromosome
soloPlots = function(){  
  for(chrom in refchr){
    writeLines(chrom)
    pngfile=paste(settings$pngbase,chrom,"png",sep=".")
    multiSNPPlot(pngfile,chrom,alt="Ref",ymin=settings$ymin,ymax=settings$ymax,xmin=settings$xmin,xmax=settings$xmax)
    pngfile=paste(settings$pngbase,chrom,"cnv.png",sep=".")
    multiSNPCNVPlot(pngfile,chrom,alt="Ref",ymin=settings$ymin,ymax=settings$ymax,xmin=settings$xmin,xmax=settings$xmax)
  }
  
  for(chrom in altchr){
    writeLines(chrom)
    pngfile=paste(settings$pngbase,chrom,"png",sep=".")
    multiSNPPlot(pngfile,chrom,alt="Alt",ymin=settings$ymin,ymax=settings$ymax,xmin=settings$xmin,xmax=settings$xmax)
    pngfile=paste(settings$pngbase,chrom,"cnv.png",sep=".")
    multiSNPCNVPlot(pngfile,chrom,alt="Alt",ymin=settings$ymin,ymax=settings$ymax,xmin=settings$xmin,xmax=settings$xmax)
    
  }
}

#3# Multiplot of all chromosomes in a single graphic
multiPlotFull = function(makepng=TRUE){
  cn = length(settings$freqcomp)
  pext = "png"
  if(settings$fdrplot){ pext = "fdr.png" }
  for(ci in 1:cn){
    (fcomp = settings$freqcomp[ci])
    comptitle = fcomp
    if(settings$fdrplot){ paste(fcomp,"FDR") }
    if(settings$fdrplot){ plotdata = fdrdb[[fcomp]] }
    else{ plotdata = combdb[[fcomp]] }
    # Reference chromosomes
    alt = "Ref"
    yi = as.integer(length(refchr) / 3)
    if(length(refchr) %% 3 > 0){ yi = yi + 1 }
    height = yi * settings$pngwidth / 6    # Default 400 per plot
    pngfile = paste(settings$pngbase,fcomp,"ref",pext,sep=".")
    if(makepng){
      writeLines(pngfile)
      png(filename=pngfile, width=settings$pngwidth * 1.5, height=height, units = "px", pointsize=settings$pointsize)
    }
    par(mar=c(2,6,2,1))
    par(mar=c(5,6,1.5,1))
    panels = 1:(yi*3)
    layout(matrix(panels,byrow=TRUE,ncol=3))
    for(chrom in refchr[1:(yi*3-4)]){
      chromSNPFreqPlot(plotdata,cnvdata,chrom,alt,comptitle,colbychrom = FALSE)
    }
    par(mar=c(5,6,1,1))
    for(chrom in refchr[(yi*3-3):length(refchr)]){
      chromSNPFreqPlot(plotdata,cnvdata,chrom,alt,comptitle,colbychrom = FALSE)
    }
    if(makepng){
      dev.off()
    }
    
    # Alt chromosomes
    alt = "Alt"
    yi = as.integer(length(altchr) / 3)
    if(length(altchr) %% 3 > 0){ yi = yi + 1 }
    height = yi * settings$pngwidth / 6    # Default 600 per plot
    # SNPFreq plot
    pngfile = paste(settings$pngbase,fcomp,"alt",pext,sep=".")
    if(makepng){
      writeLines(pngfile)
      png(filename=pngfile, width=settings$pngwidth * 1.5, height=height, units = "px", pointsize=settings$pointsize)
    }
    par(mar=c(5,6,1.5,1))
    panels = 1:(yi*3)
    layout(matrix(panels,byrow=TRUE,ncol=3))
    for(chrom in altchr[1:(yi*3-4)]){
      chromSNPFreqPlot(plotdata,cnvdata,chrom,alt,comptitle,colbychrom = FALSE)
    }
    for(chrom in altchr[(yi*3-3):length(altchr)]){
      chromSNPFreqPlot(plotdata,cnvdata,chrom,alt,comptitle,colbychrom = FALSE)
    }
    if(makepng){
      dev.off()
    }
  }  
}
if(settings$testing){
  multiPlotFull()
}

### Generate PNG Plots
#!# Add settings to specificy which plots to make and enable zooming in

zoom = settings$chrom %in% refchr ||  settings$chrom %in% altchr

if(zoom){
  writeLines(c("","Zoom plots..."))
  zoomPlot()
}else{
  if(settings$multiplot){
    writeLines(c("","Multichromosome plots..."))
    multiPlotFull()
  }else{
    writeLines(c("","Chromosome plots..."))
    soloPlots()
  }
}
  
writeLines(">>Finished")





### !!! ### OLD PLOTS TO DELETE ONCE HAPPY IT'S WORKING !!! ###

#!# Modified from above. Possibly make more flexible with data frames as input
multiSNPPlot = function(pngfile,chrom,alt="Alt",ymin=0,ymax=1.0,xmin=0,xmax=-1,colbychrom=FALSE,region=FALSE){
  png(filename=pngfile, width=settings$pngwidth, height=settings$pngheight, units = "px", pointsize=settings$pointsize)
  par(mar=c(2,6,2,1))
  panels = c(1,2,3)
  layout(matrix(panels,byrow=FALSE,nrow=3))
  chromSNPFreqPlot(galdata,cnvdata,chrom,alt,"Galactose vs Glucose",ymin,ymax,xmin,xmax,colbychrom,region)
  chromSNPFreqPlot(xyldata,cnvdata,chrom,alt,"Xylose vs Glucose",ymin,ymax,xmin,xmax,colbychrom,region)
  par(mar=c(5,6,2,1))
  chromSNPFreqPlot(popdata,cnvdata,chrom,alt,"Evolved Population P12 vs P6",0.0,ymax,xmin,xmax,colbychrom,region)
  #NB. Dropped chromCNVPlot(cnvdata,chrom) as fourth plot: not that informative in this context.
  dev.off()
}

multiSNPExpPlot = function(pngfile,chrom,alt="Alt",ymin=0,ymax=1.0,xmin=0,xmax=-1,colbychrom=FALSE,region=FALSE){
  png(filename=pngfile, width=settings$pngwidth, height=settings$pngheight, units = "px", pointsize=settings$pointsize)
  par(mar=c(2,6,2,1))
  panels = c(1,2,3)
  layout(matrix(panels,byrow=FALSE,nrow=3))
  chromSNPFreqExpPlot(galdata,cnvdata,chrom,alt,"Galactose",ymin,ymax,xmin,xmax,colbychrom,region)
  chromSNPFreqExpPlot(xyldata,cnvdata,chrom,alt,"Xylose",ymin,ymax,xmin,xmax,colbychrom,region)
  #par(mar=c(5,6,2,1))
  chromSNPFreqExpPlot(xyldata,cnvdata,chrom,alt,"Glucose",ymin,ymax,xmin,xmax,colbychrom,region,inverse=TRUE)
  #NB. Dropped chromCNVPlot(cnvdata,chrom) as fourth plot: not that informative in this context.
  dev.off()
}


makeMBGPlots = function(fileset="biallelic-11a"){
  
  (settings$basename = fileset)
  (basefile = paste("/Users/redwards/Documents/Projects/Microbiogen-Nov14/dev/2016-06-21.PileupSNPMap/",fileset,sep=""))
  (settings$pngbase = paste(basefile,".Plots/",settings$basename,sep=""))     # Directory for Plot output
  if(file.exists(paste(basefile,".Plots/",sep=""))==FALSE){
    dir.create(paste(basefile,".Plots/",sep=""))
  }
  
  (galbase = paste("/Users/redwards/Documents/Projects/Microbiogen-Nov14/dev/2016-06-21.PileupSNPMap/Gal.vs.Glu",fileset,sep="."))
  (xylbase = paste("/Users/redwards/Documents/Projects/Microbiogen-Nov14/dev/2016-06-21.PileupSNPMap/Xyl.vs.Glu",fileset,sep="."))
  (popbase = paste("/Users/redwards/Documents/Projects/Microbiogen-Nov14/dev/2016-06-21.PileupSNPMap/P12.vs.P6",fileset,sep="."))
  
  galdata = read.table( paste(galbase,".combinedsnp.tdt",sep=""), header = T, stringsAsFactors = T, sep = '\t', quote = '')
  summary(galdata)
  xyldata = read.table( paste(xylbase,".combinedsnp.tdt",sep=""), header = T, stringsAsFactors = T, sep = '\t', quote = '')
  summary(xyldata)
  popdata = read.table( paste(popbase,".combinedsnp.tdt",sep=""), header = T, stringsAsFactors = T, sep = '\t', quote = '')
  summary(popdata)
  
  settings$pcut = 0.01
  multiSNPPlot(paste(settings$pngbase,"new.exampleXV","png",sep="."),altchr[16],alt="Alt",ymin=0.0,ymax=1.0,xmin=0,xmax=400,region="Example Region:")
  multiSNPExpPlot(paste(settings$pngbase,"exp.exampleXV","png",sep="."),altchr[16],alt="Alt",ymin=0.0,ymax=1.0,xmin=0,xmax=400,region="Example Region:")
  multiSNPPlot(paste(settings$pngbase,"exampleXV","png",sep="."),altchr[16],alt="Alt",ymin=0.0,ymax=1.0,xmin=0,xmax=400,colbychrom=TRUE)
  
  for(chrom in altchr){
    pngfile=paste(settings$pngbase,chrom,"png",sep=".")
    multiSNPPlot(pngfile,chrom,alt="Alt")
  }
  
  for(chrom in refchr){
    pngfile=paste(settings$pngbase,chrom,"png",sep=".")
    multiSNPPlot(pngfile,chrom,alt="Ref")
  }
  
}


if(settings$testing){
  

  #### Gal, Xyl and P12/P6 Data !
  galbase = "/Volumes/bioinf/aperezbercoff/Microbiogen-Nov14/dev/2016-06-21.PileupSNPMap/Gal.vs.Glu"
  xylbase = "/Volumes/bioinf/aperezbercoff/Microbiogen-Nov14/dev/2016-06-21.PileupSNPMap/Xyl.vs.Glu"
  popbase = "/Volumes/bioinf/aperezbercoff/Microbiogen-Nov14/dev/2016-06-21.PileupSNPMap/P12.v.P6"
  
  galdata = read.table( paste(galbase,".combinedsnp.tdt",sep=""), header = T, stringsAsFactors = T, sep = '\t', quote = '')
  summary(galdata)
  xyldata = read.table( paste(xylbase,".combinedsnp.tdt",sep=""), header = T, stringsAsFactors = T, sep = '\t', quote = '')
  summary(xyldata)
  popdata = read.table( paste(popbase,".combinedsnp.tdt",sep=""), header = T, stringsAsFactors = T, sep = '\t', quote = '')
  summary(popdata)
  
  
  
  ### Multipanel Plot ##
  
  
  ### ~ Generate multipanel plots for each assembled contig ~~~~~~~~~~~~~ ###
  for(chrom in altchr){
    pngfile=paste(settings$pngbase,chrom,"png",sep=".")
    #CairoPNG(filename=pngfile, width=pwidth, height=pheight, pointsize=12)
    png(filename=pngfile, width=settings$pngwidth, height=settings$pngheight, units = "px", pointsize=settings$pointsize)
    par(mar=c(2,6,2,1))
    #panels = c(1,1,1,2,2,2,3,3,1,1,1,2,2,2,4,4,1,1,1,2,2,2,5,5,1,1,1,2,2,2,6,6)
    #layout(matrix(panels,byrow=FALSE,nrow=8))
    panels = c(1,2,3,4)
    layout(matrix(panels,byrow=FALSE,nrow=4))
    chromSNPFreqPlot(galdata,cnvdata,chrom,alt="Alt",comparison="Gal vs Glu")
    chromSNPFreqPlot(xyldata,cnvdata,chrom,alt="Alt",comparison="Xyl vs Glu")
    chromSNPFreqPlot(popdata,cnvdata,chrom,alt="Alt",comparison="P12 vs P6")
    par(mar=c(5,6,1,1))
    chromCNVPlot(cnvdata,chrom)
    dev.off()
  }
  
  for(chrom in refchr){
    pngfile=paste(settings$pngbase,chrom,"png",sep=".")
    #CairoPNG(filename=pngfile, width=pwidth, height=pheight, pointsize=12)
    png(filename=pngfile, width=settings$pngwidth, height=settings$pngheight, units = "px", pointsize=settings$pointsize)
    par(mar=c(2,6,2,1))
    #panels = c(1,1,1,2,2,2,3,3,1,1,1,2,2,2,4,4,1,1,1,2,2,2,5,5,1,1,1,2,2,2,6,6)
    #layout(matrix(panels,byrow=FALSE,nrow=8))
    panels = c(1,2,3,4)
    layout(matrix(panels,byrow=FALSE,nrow=4))
    chromSNPFreqPlot(galdata,cnvdata,chrom,alt="Ref",comparison="Gal vs Glu")
    chromSNPFreqPlot(xyldata,cnvdata,chrom,alt="Ref",comparison="Xyl vs Glu")
    chromSNPFreqPlot(popdata,cnvdata,chrom,alt="Ref",comparison="P12 vs P6")
    par(mar=c(5,6,1,1))
    chromCNVPlot(cnvdata,chrom)
    dev.off()
  }
  
  
  
  
  chrom = altchr[16]
  
  settings$pcut = 0.001
  multiSNPPlot(paste(settings$pngbase,"exampleXV","png",sep="."),altchr[16],alt="Alt",ymin=0.0,ymax=1.0,xmin=0,xmax=400)
  
  for(chrom in altchr){
    pngfile=paste(settings$pngbase,chrom,"png",sep=".")
    multiSNPPlot(pngfile,chrom,alt="Alt")
  }
  
  for(chrom in refchr){
    pngfile=paste(settings$pngbase,chrom,"png",sep=".")
    multiSNPPlot(pngfile,chrom,alt="Ref")
  }
  
  
  
  
  
  ### >>> Biallelic Run >>>>>
  
  #### Gal, Xyl and P12/P6 Data !
  galbase = "/Users/redwards/Documents/Projects/Microbiogen-Nov14/dev/2016-06-21.PileupSNPMap/Gal.vs.Glu.biallelic"
  xylbase = "/Users/redwards/Documents/Projects/Microbiogen-Nov14/dev/2016-06-21.PileupSNPMap/Xyl.vs.Glu.biallelic"
  popbase = "/Users/redwards/Documents/Projects/Microbiogen-Nov14/dev/2016-06-21.PileupSNPMap/P12.vs.P6.biallelic"
  
  
  
  
  ### >> 11a Alleles Only >>
  
  makeMBGPlots()
  
  makeMBGPlots("biallelic-11a-all")

}


